{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Crossovers in genetic algorithms\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "**Crossover**, also known as recombination, is the process of combining two parent solutions to generate new offspring. This results in two offspring, each carrying some genetic information from both parents. The idea is to exchange some parts of the parents' strings to create a new string that may inherit some good traits from both parents, and it is analogous to the crossover that happens during sexual reproduction in biology.\n",
    "\n",
    "In evolutionary computation, various algorithms may use distinct data structures to store genetic information (see [genetic algorithms](genetic_algorithms.ipynb)). Each genetic representation can be combined with different crossover operators. Bit arrays, vectors of real numbers, or trees are some typical data structures that can be recombined with crossover. Note that the same individual may be selected as both parents, and one individual may take part in producing more than one offspring. \n",
    "\n",
    "**Main categories** of crossover operators are:  <br>\n",
    "(1) <span style=\"color:#EE8877\">Asexual:</span> an offspring is generated from one parent, <br>\n",
    "(2) <span style=\"color:#77EFE5\">Sexual:</span> two parents produce one or more offspring, and <br>\n",
    "(3) <span style=\"color:#8799EE\">Multi-recombination:</span> more than two parents are used to produce one or more offspring."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/crossover.png\"  alt='missing' width=\"400\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Contents**\n",
    "\n",
    "<a href='#chrom-def'>01. Chromosome Definition</a>  \n",
    " \n",
    "<a href='#one-point'>02. One-point Crossover</a>\n",
    "\n",
    "<a href='#multi-point'>03. Multi-point Crossover</a>\n",
    "\n",
    "<a href='#uniform'>04. Uniform Crossover</a>\n",
    "\n",
    "<a href='#multi-parent'>05. Multi-parent Crossover</a>\n",
    "\n",
    "<a href='#order'>06. Order Crossover (OX)</a>\n",
    "\n",
    "<a href='#partially-mapped'>07. Partially Mapped Crossover (PMX)</a>\n",
    "\n",
    "<a href='#cycle'>08. Cycle Crossover (CX)</a>\n",
    "\n",
    "<a href='#intermediate'>09. Heuristic Crossover (HC) or Intermediate Crossover (IC)</a>\n",
    "\n",
    "<a href='#shuffle'>10. Shuffle Crossover (SX)</a>\n",
    "\n",
    "<a href='#reduce-surrogate'>11. Reduce Surrogate Crossover (RSX)</a>\n",
    "\n",
    "<a href='#average'>12. Average Crossover (AX)</a>\n",
    "\n",
    "<a href='#discrete'>13. Discrete Crossover (DC)</a>\n",
    "\n",
    "<a href='#flat'>14. Flat Crossover (FC)</a>\n",
    "\n",
    "<a href='#multivariate'>15. Multivariate Crossover</a>\n",
    "\n",
    "<a href='#count-preserving'>16. Count Preserving Crossover (CPC)</a>\n",
    "\n",
    "<a href='#modified-order'>17. Modified Order Crossover (MOC)</a>\n",
    "\n",
    "<a href='#random-respectful'>18. Random Respectful Crossover (RRC)</a>\n",
    "\n",
    "<a href='#masked'>19. Masked Crossover</a>\n",
    "\n",
    "<a href='#1-bit-adaptive'>20. 1-Bit Adaptive Crossover (1BX)</a>\n",
    "\n",
    "<a href='#modified-partially-mapped'>21. Modified Partially Mapped Crossover (MPMX)</a>\n",
    "\n",
    "<a href='#order-based'>22. Order-Based Crossover (OBX)</a>\n",
    "\n",
    "<a href='#position-based'>23. Position-Based Crossover (POS)</a>\n",
    "\n",
    "<a href='#voting-recombination'>24. Voting Recombination Crossover (VR)</a>\n",
    "\n",
    "<a href='#maximal-preservation'>25. Maximal Preservation Crossover (MPX)</a>\n",
    "\n",
    "<a href='#position'>26. Position Crossover (PX)</a>\n",
    "\n",
    "<a href='#homologous'>27. Homologous Crossover (HX)</a>\n",
    "\n",
    "<a href='#complete-subtour-exchange'>28. Complete Subtour Exchange Crossover (CSEX)</a>\n",
    "\n",
    "<a href='#edge-recombination'>29. Edge Recombination Crossover (ER)</a>\n",
    "\n",
    "<a href='#alternate-edges'>30. Alternate Edges Crossover (AEX)</a>\n",
    "\n",
    "<a href='#half-uniform'>31. Half Uniform Crossover (HUX)</a>\n",
    "\n",
    "<a href='#highly-disruptive'>32. Highly Disruptive Crossover (HDX)</a>\n",
    "\n",
    "<a href='#blend'>33. Blend Crossover (BLX)</a>\n",
    "\n",
    "<a href='#linear-recombination'>34. Linear Recombination Crossover</a>\n",
    "\n",
    "<a href='#directional-heuristic'>35. Directional Heuristic Crossover</a>\n",
    "\n",
    "<a href='#geometrical'>36. Geometrical Crossover</a>\n",
    "\n",
    "<a href='#simulated-binary'>37. Simulated Binary Crossover (SBX)</a>\n",
    "\n",
    "<a href='#hill-climbing'>38. Hill-Climbing Crossover</a>\n",
    "\n",
    "<a href='#arithmetic'>39. Arithmetic Crossover</a>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='chrom-def'>Chromosome Definition</a>\n",
    "\n",
    "First, we need to implement the chromosome class, a function to calculate the fitness of individuals (chromosomes) in the population, and some helper functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "class Chromosome():\n",
    "    \"\"\"\n",
    "    Description of class `Chromosome`:\n",
    "    This class represents a simple chromosome. In the method describe, a simple description\n",
    "    of the chromosome is provided, when it is called. \n",
    "    \"\"\"\n",
    "    def __init__(self, genes, id_=None, fitness=-1):\n",
    "        self.id_ = id_\n",
    "        self.genes = genes\n",
    "        self.fitness = fitness       \n",
    "       \n",
    "    def describe(self): \n",
    "        \"\"\"\n",
    "        Prints the ID, fitness, and genes\n",
    "        \"\"\"\n",
    "        print(f\"ID=#{self.id_}, Fitness={self.fitness}, \\nGenes=\\n{self.genes}\")\n",
    " \n",
    "    def get_chrom_length(self): \n",
    "        \"\"\"\n",
    "        Returns the length of `self.genes`\n",
    "        \"\"\"\n",
    "        return len(self.genes)\n",
    "\n",
    "def find_chrom_type(chrom):\n",
    "    \"\"\"\n",
    "    This function, takes a chromosome and returns its type (binary, integer, \n",
    "    or a floating-point chromosome).\n",
    "    chrom: The chromoseme which its type is calculated and returned.\n",
    "    \"\"\" \n",
    "    # if a floating-point individual\n",
    "    if np.issubdtype(chrom.dtype, np.floating):\n",
    "        return \"float\"\n",
    "    # if a binary individual\n",
    "    elif np.array_equal(chrom, chrom.astype(bool)):\n",
    "        return \"binary\"\n",
    "    # if an integer individual\n",
    "    elif np.issubdtype(chrom.dtype, np.integer):\n",
    "        return \"integer\"\n",
    "    \n",
    "    return \"binary\"\n",
    "\n",
    "# Note that calculating the fitness of a chromosome depends on the problem at hand,\n",
    "# and what you see in this function is just for understanding the concept.\n",
    "def fitness_function(chrom): \n",
    "    \"\"\"\n",
    "    This function, takes a chromosome and returns a value as its fitness.\n",
    "    chrom: The chromoseme which its fitness is calculated and returned.\n",
    "    \"\"\"  \n",
    "    chrom_type = find_chrom_type(chrom)\n",
    "    # if we have a binary chromosome (count the number of 1s)\n",
    "    if chrom_type == \"binary\":\n",
    "        return np.count_nonzero(chrom == 1)\n",
    "    # if we have an integer chromosome (count values between 0 and 4)\n",
    "    elif chrom_type == \"integer\":\n",
    "        return np.count_nonzero((chrom >= 0) & (chrom <= 4))\n",
    "    # if we have a floating-point chromosome (count values between 0 and 0.5) \n",
    "    elif chrom_type == \"float\":\n",
    "        return np.count_nonzero((chrom >= 0) & (chrom <= 0.5))\n",
    "    \n",
    "    return -1\n",
    "\n",
    "\n",
    "def sorted_random_numbers(point_nums, interval):\n",
    "    \"\"\"\n",
    "    Returns several sorted random integers in an interval.\n",
    "    point_nums: number of points to be selected.\n",
    "    interval: the interval which the numbers are selected.\n",
    "    \"\"\"    \n",
    "    points = []\n",
    "    for _ in range(point_nums):\n",
    "        num = np.random.randint(1, abs(interval))\n",
    "        while num in points:\n",
    "            num = np.random.randint(1, abs(interval))\n",
    "        points.append(num)\n",
    "    return sorted(points)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='one-point'>One-point Crossover</a>\n",
    "In one-point crossover, two chromosomes as parents are selected and two offspring are produced. One point in the parent chromosomes is selected randomly, and the genes of the first part of child-one will be filled by the parent-one, and the second part of the child-one will be filled by the second part of the parent-two. For child-2, the first part is filled by parent-two, and the second part is filled by the parent-one."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/one_point_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=0, \n",
      "Genes=\n",
      "[0 0 0 0 0 0 0 0 0 0 0]\n",
      "ID=#1, Fitness=11, \n",
      "Genes=\n",
      "[1 1 1 1 1 1 1 1 1 1 1]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=3, \n",
      "Genes=\n",
      "[0 0 0 0 0 0 0 0 1 1 1]\n",
      "ID=#1, Fitness=8, \n",
      "Genes=\n",
      "[1 1 1 1 1 1 1 1 0 0 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def one_point_crossover(parent_one, parent_two, x_point, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs One-point crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    x_point: The crossover point\n",
    "    pc: The probability of crossover \n",
    "    \"\"\"  \n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0]*chrom_length),id_=0,fitness = 125.2)  \n",
    "    child_two = Chromosome(genes= np.array([0]*chrom_length),id_=1,fitness = 125.2)\n",
    "    point = np.random.randint(0, abs(chrom_length))\n",
    "    if x_point != 0: point = x_point\n",
    "    # point = 4\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        for i in range(chrom_length):\n",
    "            if i<point:\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "            else:\n",
    "                child_one.genes[i] = parent_two.genes[i]\n",
    "                child_two.genes[i] = parent_one.genes[i]\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "# parents\n",
    "chrom0 = np.array([0,0,0,0,0,0,0,0,0,0,0])\n",
    "chrom1 = np.array([1,1,1,1,1,1,1,1,1,1,1])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "\n",
    "CROSS = one_point_crossover(parent_one, parent_two, 0, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='multi-point'>Multi-point Crossover</a>\n",
    "\n",
    "This crossover is an extension of one-point crossover. Several points in the parent chromosomes are randomly selected, and the genes from the parents will be transferred to children in the same order, but the parent who is transferring the genes will change at each point."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/multi_point_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=0, \n",
      "Genes=\n",
      "[0 0 0 0 0 0 0 0 0 0 0]\n",
      "ID=#1, Fitness=11, \n",
      "Genes=\n",
      "[1 1 1 1 1 1 1 1 1 1 1]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[0 0 0 1 1 0 0 0 1 1 0]\n",
      "ID=#1, Fitness=7, \n",
      "Genes=\n",
      "[1 1 1 0 0 1 1 1 0 0 1]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def multi_point_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Multi-point crossover on them by selecting \n",
    "    several points randomly, as crossover points. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover \n",
    "    \"\"\"  \n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([0]*chrom_length),id_=1,fitness = 125.2)\n",
    "    point_nums = np.random.randint(1, abs(chrom_length)) #a random number for the number of points\n",
    "    points = sorted_random_numbers(point_nums,chrom_length)\n",
    "    count = 0\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        prev_index=0\n",
    "        for _ in range(len(points)):\n",
    "            if count%2==0:\n",
    "                for i in range(prev_index,chrom_length,1):\n",
    "                    if i<points[count]:\n",
    "                        child_one.genes[i] = parent_one.genes[i]\n",
    "                        child_two.genes[i] = parent_two.genes[i]\n",
    "                prev_index = points[count]\n",
    "                count +=1\n",
    "            elif count%2!=0:\n",
    "                for i in range(prev_index,chrom_length,1):\n",
    "                    if i<points[count]:\n",
    "                        child_one.genes[i] = parent_two.genes[i]\n",
    "                        child_two.genes[i] = parent_one.genes[i]\n",
    "                prev_index = points[count]\n",
    "                count +=1          \n",
    "        # for the last part\n",
    "        if count%2==0:\n",
    "            for i in range(points[len(points)-1],chrom_length,1):         \n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "        elif count%2!=0:\n",
    "            for i in range(points[len(points)-1],chrom_length,1):\n",
    "                child_one.genes[i] = parent_two.genes[i]\n",
    "                child_two.genes[i] = parent_one.genes[i]\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "    \n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "# parents\n",
    "chrom0 = np.array([0,0,0,0,0,0,0,0,0,0,0])\n",
    "chrom1 = np.array([1,1,1,1,1,1,1,1,1,1,1])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "\n",
    "CROSS = multi_point_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='uniform'>Uniform Crossover</a>\n",
    "\n",
    "In uniform crossover, a random number between 0 and 1 is produced for each gene of the parent chromosome. If that random number is less than 0.5, then the gene from parent-one will be transferred to child-one; otherwise, the gene will come from parent-two.\n",
    "The same thing will happen to child two, but the condition is reversed. Which means that whatever gene is transferred to child-one from one of the parents, the gene from the other parent in that position is transferred to child-two."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/uniform_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=0, \n",
      "Genes=\n",
      "[0 0 0 0 0 0 0 0 0 0 0]\n",
      "ID=#1, Fitness=11, \n",
      "Genes=\n",
      "[1 1 1 1 1 1 1 1 1 1 1]\n",
      "\n",
      " Mask Vector\n",
      "=================================================\n",
      "[0 1 0 1 0 0 0 1 0 0 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=8, \n",
      "Genes=\n",
      "[1 0 1 0 1 1 1 0 1 1 1]\n",
      "ID=#1, Fitness=3, \n",
      "Genes=\n",
      "[0 1 0 1 0 0 0 1 0 0 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def claculate_mask(chrom_size, px):\n",
    "    \"\"\"\n",
    "    This method, performs the mask calculation for the Uniform crossover.\n",
    "    chrom_size: The mask has the same size of the chromosome.\n",
    "    px: A control parameter for the number of ones in the mask vector.\n",
    "    \"\"\"  \n",
    "    mask = [0]*chrom_size\n",
    "    for i in range(chrom_size):\n",
    "        prob = np.random.rand()\n",
    "        if prob <= px:\n",
    "            mask[i]=1\n",
    "    return np.array(mask)\n",
    "            \n",
    "def uniform_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Uniform crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0]*chrom_length),id_=0,fitness = 125.2) \n",
    "    child_two = Chromosome(genes= np.array([0]*chrom_length),id_=1,fitness = 125.2)\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        mask = claculate_mask(chrom_length, np.random.uniform(low=0.2, high=0.85))\n",
    "        print(\"\\n Mask Vector\")\n",
    "        print(\"=================================================\")\n",
    "        print(mask)\n",
    "        for i in range(chrom_length):\n",
    "            if mask[i]==1:\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "            else:\n",
    "                child_one.genes[i] = parent_two.genes[i]\n",
    "                child_two.genes[i] = parent_one.genes[i]\n",
    "           \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "# parents\n",
    "chrom0 = np.array([0,0,0,0,0,0,0,0,0,0,0])\n",
    "chrom1 = np.array([1,1,1,1,1,1,1,1,1,1,1])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = uniform_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='multi-parent'>Multi-parent Crossover</a>\n",
    "\n",
    "In multi-parent crossover, several parents from the population are selected and one child is produced. Each gene of the offspring is from one of its parents. For instance, if we consider six chromosomes as the parents, the offspring will be something like this."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/multi_parent_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=0, \n",
      "Genes=\n",
      "[0 0 0 0 0 0]\n",
      "ID=#1, Fitness=6, \n",
      "Genes=\n",
      "[1 1 1 1 1 1]\n",
      "ID=#2, Fitness=6, \n",
      "Genes=\n",
      "[2 2 2 2 2 2]\n",
      "ID=#3, Fitness=6, \n",
      "Genes=\n",
      "[3 3 3 3 3 3]\n",
      "ID=#4, Fitness=6, \n",
      "Genes=\n",
      "[4 4 4 4 4 4]\n",
      "ID=#5, Fitness=0, \n",
      "Genes=\n",
      "[5 5 5 5 5 5]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 5]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy         \n",
    "\n",
    "def multi_parent_crossover(parents, pc):\n",
    "    \"\"\"\n",
    "    This function takes several parents, and performs Multi-parent crossover on them. \n",
    "    In this method, number of parents is equal to the length of the chromosome.\n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parents[0])\n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    for i in range(chrom_length):\n",
    "        Chromosome.describe(parents[i])\n",
    "\n",
    "    child_one = Chromosome(genes= np.array([0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        for i in range (chrom_length):\n",
    "            child_one.genes[i]=parents[i].genes[i]           \n",
    "       \n",
    "    else:  # if pc is less than random number then the children will be one of the parents involved\n",
    "        child_one = deepcopy(parents[np.random.randint(0, chrom_length)])\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "        \n",
    "    return child_one\n",
    "\n",
    "# parents\n",
    "parents = []\n",
    "genes0 = np.array([0, 0, 0, 0, 0, 0])\n",
    "genes1 = np.array([1, 1, 1, 1, 1, 1])\n",
    "genes2 = np.array([2, 2, 2, 2, 2, 2])\n",
    "genes3 = np.array([3, 3, 3, 3, 3, 3])\n",
    "genes4 = np.array([4, 4, 4, 4, 4, 4])\n",
    "genes5 = np.array([5, 5, 5, 5, 5, 5])\n",
    "\n",
    "parents.append(Chromosome(genes= np.array(genes0), id_=0, fitness = fitness_function(genes0)))\n",
    "parents.append(Chromosome(genes= np.array(genes1), id_=1, fitness = fitness_function(genes1)))\n",
    "parents.append(Chromosome(genes= np.array(genes2), id_=2, fitness = fitness_function(genes2)))\n",
    "parents.append(Chromosome(genes= np.array(genes3), id_=3, fitness = fitness_function(genes3)))\n",
    "parents.append(Chromosome(genes= np.array(genes4), id_=4, fitness = fitness_function(genes4)))\n",
    "parents.append(Chromosome(genes= np.array(genes5), id_=5, fitness = fitness_function(genes5)))\n",
    "\n",
    "\n",
    "CROSS = multi_parent_crossover(parents, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    \n",
    "    Chromosome.describe(CROSS)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='order'>Order Crossover (OX)</a>\n",
    "\n",
    "In order crossover, two points are selected randomly in the parents. Then the genes between these two points from parent-one is directly transferred to child-one and from parent-two to child-two. After that, for child-one we start from the second point of parent-two and we transfer to child-one what is not already transferred to child-one. We do the same thing for child-two with parent-one."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/order_cross.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[4 5 2 1 8 0 7 6 9 3]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 7 2 3 4 5 6 9 1 8]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[5 6 2 1 8 0 7 9 3 4]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy  \n",
    "\n",
    "def order_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Order crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)  \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "    point1 = np.random.randint(1, int(chrom_length/2))\n",
    "    point2 = np.random.randint(int(chrom_length/2), chrom_length)\n",
    "    \n",
    "    trans_elements1 = [] #elements transfered to child1 from parent1\n",
    "    trans_elements2 = [] #elements transfered to child2 from parent2\n",
    "    list1 = [] #elements to be transfered to child1 from parent2 \n",
    "    list2 = [] #elements to be transfered to child2 from parent1\n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        for i in range(chrom_length):  # transfer genes\n",
    "            if i >= point1 and i < point2:\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "                trans_elements1.append(child_one.genes[i])\n",
    "                trans_elements2.append(child_two.genes[i])\n",
    "        for i in range(point2,chrom_length):\n",
    "            list1.append(parent_two.genes[i]) #for c1 from p2\n",
    "            list2.append(parent_one.genes[i]) #for c2 from p1\n",
    "        for i in range(point2):\n",
    "            list1.append(parent_two.genes[i]) \n",
    "            list2.append(parent_one.genes[i]) \n",
    "        delete1 = [] #index of the elements that should be deleted from list1\n",
    "        delete2 = [] #index of the elements that should be deleted from list2\n",
    "        for i in range(chrom_length):\n",
    "            if list1[i] in trans_elements1: delete1.append(i)\n",
    "            if list2[i] in trans_elements2: delete2.append(i)\n",
    "        count = 0\n",
    "        for i in range(len(delete1)): #deleting the redundant elements\n",
    "            list1.pop(delete1[i]-count)\n",
    "            list2.pop(delete2[i]-count)\n",
    "            count +=1\n",
    "        count = 0\n",
    "        for i in range(point2,chrom_length):\n",
    "            child_one.genes[i] = list1[count]\n",
    "            child_two.genes[i] = list2[count]\n",
    "            count +=1 \n",
    "            \n",
    "        for i in range(point1):\n",
    "            child_one.genes[i] = list1[count]\n",
    "            child_two.genes[i] = list2[count]\n",
    "            count +=1\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "    \n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "# parents\n",
    "chrom0 = np.array([0,1,2,3,4,5,6,7,8,9])\n",
    "chrom1 = np.array([4,5,2,1,8,0,7,6,9,3])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = order_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='partially-mapped'>Partially Mapped Crossover (PMX)</a>\n",
    "\n",
    "In this crossover, two points are selected randomly in the parents. Then the genes between these two points from parent-one are directly transferred to child-one and from parent-two to child-two. After that, for child-one, we look in parent-two and transfer all the genes that can be transferred (not already exist in the middle section of child-one), directly to child-one. And we do the same thing for child-two but with parent-one.\n",
    "\n",
    "For the last step, for child-one, we go through the middle section of parent-two and find an element that does not already exist in child-one. Then, we search the value that exists in its corresponding index in child-one in parent-two. If the corresponding index of the searched value in parent-two is empty in child-one, we put the element in that spot. Else, we keep trying until we find an empty spot. \n",
    "\n",
    "In the following example, after the first and the second step, we have these for child-one and child-two:\n",
    "\n",
    "`child-one = [0 1 -1 3 6 2 -1 7 8 9]` <br><br>\n",
    "`child-two = [8 -1 7 3 4 5 -1 1 9 0]`\n",
    "\n",
    "where -1 denotes an empty spot that needs to be filled. For finishing child-one, we go through parent-two and **the first element that we see that already doesn't exist in child-one is 4.** We notice that in the same index, we have 6 in child-one. We search through parent-two for the element 6 and we find it in the 6th index of parent-two (counting from 0 from left). We look at the corresponding position in child-one and we see that it is empty (denoted by -1). So we put 4 in that position. <br> \n",
    "Similarly, **the next element in parent-two that doesn't already exist in child-one is 5.** In the corresponding position in child-one, we have 2. We search for 2 in parent-two and we find it in the 3rd index from the left. We check the same position in child-one and we notice that it is empty, so 5 into it. As a result, child-one would be:\n",
    "\n",
    "`child-one = [0 1 5 3 6 2 4 7 8 9]` <br>\n",
    "\n",
    "We have the same process for filling the child-two, but this time with parent-one.\n",
    "\n",
    "<span style=\"color:#55CABB\">PMX is one of the most commonly used crossover operators for permutation encoded chromosomes. It is widely used in various fields, including engineering, finance, and game designing, to solve complex problems.</span>\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/pmx.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[4 5 2 1 8 0 7 6 9 3]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[8 5 2 1 4 0 7 6 9 3]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 8 5 6 7 4 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy \n",
    "\n",
    "def partially_mapped_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Partially-mapped-crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    # select two random points\n",
    "    point1 = np.random.randint(1, abs(chrom_length/2))\n",
    "    point2 = np.random.randint(abs(chrom_length/2), chrom_length - 1)\n",
    "    # point1 = 3\n",
    "    # point2 = 8\n",
    "\n",
    "    NotSeenList1 = []\n",
    "    NotSeenList2 = []\n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "\n",
    "        child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "        child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "\n",
    "        for i in range(chrom_length):  # transfer genes\n",
    "            if i >= point1 and i < point2:\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "        for i in range(chrom_length):\n",
    "            if i < point1 or i >= point2:\n",
    "                # elements that did not already transfered to child-one\n",
    "                NotSeenList1.append(parent_one.genes[i])\n",
    "                # elements that did not already transfered to child-two\n",
    "                NotSeenList2.append(parent_two.genes[i])\n",
    "\n",
    "        for i in range(chrom_length):  # First child\n",
    "            if(child_one.genes[i] != -1 and parent_two.genes[i] in NotSeenList1):\n",
    "                ans = parent_two.genes[i]\n",
    "                # keep searching until you find an empty spot\n",
    "                while child_one.genes[i] != -1: \n",
    "                    i = parent_two.genes.tolist().index(child_one.genes[i])\n",
    "                child_one.genes[i] = ans\n",
    "\n",
    "        for i in range(chrom_length):\n",
    "            if parent_two.genes[i] in NotSeenList1:\n",
    "                if(child_one.genes[i] == -1):\n",
    "                    child_one.genes[i] = parent_two.genes[i]\n",
    "\n",
    "        for i in range(chrom_length):  # Second child\n",
    "            if(child_two.genes[i] != -1 and parent_one.genes[i] in NotSeenList2):\n",
    "                tmp = parent_one.genes[i]\n",
    "                # keep searching until you find an empty spot\n",
    "                while child_two.genes[i] != -1:\n",
    "                    i = parent_one.genes.tolist().index(child_two.genes[i])\n",
    "                child_two.genes[i] = tmp\n",
    "\n",
    "        for i in range(chrom_length):\n",
    "            if parent_one.genes[i] in NotSeenList2:\n",
    "                if(child_two.genes[i] == -1):\n",
    "                    child_two.genes[i] = parent_one.genes[i]\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0,1,2,3,4,5,6,7,8,9])\n",
    "chrom1 = np.array([4,5,2,1,8,0,7,6,9,3])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = partially_mapped_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='cycle'>Cycle Crossover (CX)</a>\n",
    "\n",
    "The cycle crossover operator identifies a number of so-called cycles between two parent chromosomes. Then, to form child 1, cycle one is copied from parent 1, cycle 2 from parent 2, cycle 3 from parent 1, and so on. Consider the following example. We can finish the children in 3 cycles."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/cycle_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Cycle 1** <br>\n",
    "`Values 8 9 0`:  We start with the first value in Parent 1 and drop down to the same position in Parent 2. 8 Goes to 0. Then, we look for 0 in Parent 1 and find it at the 10th position where we drop down to 9. Again, we look for this value in Parent 1 and find it in the 9th position and drop down to 8. Since we started with 8, we've completed our cycle.\n",
    "\n",
    "**Cycle 2** <br>\n",
    "`Values 4 1 7 2 5 6`: We start with 4 and drop down to 1. 1 is found in the 8th position in Parent 1 and we drop down to 7. 7 Drops down to 2, 2 Drops down to 5, 5 drops down to 6, and 6 drops down to 4 - Our cycle is complete.\n",
    "\n",
    "**Cycle 3** <br>\n",
    "`Value 3`: The only possible cycle left is of length 1 and contains the value 3.\n",
    "\n",
    "Filling in the offspring: <br>\n",
    "`Copy Cycle 1:` Cycle 1 values from Parent 1 are copied to Child 1, and values from Parent 2 will be copied to Child 2. Cycle 2 will by different. <br>\n",
    "`Copy Cycle 2:` Cycle 2 values from Parent 1 will be copied to Child 2, and values from Parent 1 will be copied to Child 1. <br>\n",
    "`Copy Cycle 3:` Cycle 3 is like Cycle 1, Parent 1 goes to Child 1, Parent 2 goes to Child 2.\n",
    "\n",
    "This alternating process goes on until the children are finished. <span style=\"color:#55CABB\">This crossover is used for problems such as the traveling salesman, to find the shortest possible route, over generations.</span>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[4 5 2 1 8 0 7 6 9 3]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[4 5 2 1 8 0 7 6 9 3]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy \n",
    "\n",
    "def cycle_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Cycle crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"\n",
    "\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    child_one = Chromosome(genes=np.array([-1] * chrom_length), id_=0, fitness=125.2)\n",
    "    child_two = Chromosome(genes=np.array([-1] * chrom_length), id_=1, fitness=125.2)\n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        p1_copy = parent_one.genes.tolist()\n",
    "        p2_copy = parent_two.genes.tolist()\n",
    "        swap = True\n",
    "        count = 0\n",
    "        pos = 0\n",
    "\n",
    "        while True:\n",
    "            if count > chrom_length:\n",
    "                break\n",
    "            for i in range(chrom_length):\n",
    "                if child_one.genes[i] == -1:\n",
    "                    pos = i\n",
    "                    break\n",
    "\n",
    "            if swap:\n",
    "                while True:\n",
    "                    child_one.genes[pos] = parent_one.genes[pos]\n",
    "                    count += 1\n",
    "                    pos = parent_two.genes.tolist().index(parent_one.genes[pos])\n",
    "                    if p1_copy[pos] == -1:\n",
    "                        swap = False\n",
    "                        break\n",
    "                    p1_copy[pos] = -1\n",
    "            else:\n",
    "                while True:\n",
    "                    child_one.genes[pos] = parent_two.genes[pos]\n",
    "                    count += 1\n",
    "                    pos = parent_one.genes.tolist().index(parent_two.genes[pos])\n",
    "                    if p2_copy[pos] == -1:\n",
    "                        swap = True\n",
    "                        break\n",
    "                    p2_copy[pos] = -1\n",
    "\n",
    "        for i in range(chrom_length): #for the second child\n",
    "            if child_one.genes[i] == parent_one.genes[i]:\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "            else:\n",
    "                child_two.genes[i] = parent_one.genes[i]\n",
    "\n",
    "        for i in range(chrom_length): #Special mode\n",
    "            if child_one.genes[i] == -1:\n",
    "                if p1_copy[i] == -1: #it means that the ith gene from p1 has been already transfered\n",
    "                    child_one.genes[i] = parent_two.genes[i]\n",
    "                else:\n",
    "                    child_one.genes[i] = parent_one.genes[i]\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0,1,2,3,4,5,6,7,8,9])\n",
    "chrom1 = np.array([4,5,2,1,8,0,7,6,9,3])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = cycle_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='intermediate'>Heuristic Crossover (HC) or Intermediate Crossover (IC)</a>\n",
    "\n",
    "In this crossover two parents are selected from the population and one offspring is produced. We first create a random number between 0 and 1. Then the genes of the child chromosome will be calculated as follows: <br>\n",
    "\n",
    "`child_one[i] = parent_one[i] + a*(parent_two[i] - parent_one[i])`\n",
    "\n",
    "where variable `a` is a random number between 0 and 1."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/heuristic_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.121 0.152 0.231 0.143 0.732 0.315 0.434 0.633]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[0.765 0.168 0.914 0.435 0.893 0.332 0.981 0.194]\n",
      "0.5924962257560331\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=3, \n",
      "Genes=\n",
      "[0.503 0.161 0.636 0.316 0.827 0.325 0.758 0.893]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def heuristic_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Heuristic crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0.0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    # child_two = chromosome(genes= np.array([-1]*chrom_length),id=1,fitness = 125.2)\n",
    "    a = np.random.uniform(low=0, high=1)\n",
    "    print(a)\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        for i in range(chrom_length):\n",
    "            child_one.genes[i] = round(parent_one.genes[i] + a*np.abs(parent_two.genes[i] - parent_one.genes[i]), 3)\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        \n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "\n",
    "    return child_one\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0.121, 0.152, 0.231, 0.143, 0.732, 0.315, 0.434, 0.633])\n",
    "chrom1 = np.array([0.765, 0.168, 0.914, 0.435, 0.893, 0.332, 0.981, 0.194])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = heuristic_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(CROSS)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='shuffle'>Shuffle Crossover (SX)</a>\n",
    "\n",
    "In this crossover, we select two parents from the population and two offspring are produced. We choose several random positions which are the same in both parents. Then we shuffle the genes in these positions in the same way in both parents. After that, we do one-point crossover on the parents and finally unshuffle the genes in the offspring that we shuffled earlier."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/shuffle_crossover.png\"  alt='missing' width=\"700\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 1 1 0 1 0 0 1 0]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 1 1 0]\n",
      "\n",
      "Parents after performing shuffles:\n",
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 0 1 0 1 0 1 1 0]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[1 1 0 0 1 0 0 1 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 1 1 0]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[1 1 1 0 1 0 0 1 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy \n",
    "\n",
    "def shuffle_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Shuffle crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")    \n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "    points = sorted_random_numbers(4,chrom_length)\n",
    "    # print(points)\n",
    "  \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        # shuffling\n",
    "        parent_one.genes[points[0]], parent_one.genes[points[2]] = parent_one.genes[points[2]], parent_one.genes[points[0]] #shuffle the elements of p1\n",
    "        parent_one.genes[points[1]], parent_one.genes[points[3]] = parent_one.genes[points[3]], parent_one.genes[points[1]]\n",
    "\n",
    "        parent_two.genes[points[0]], parent_two.genes[points[2]] = parent_two.genes[points[2]], parent_two.genes[points[0]] #shuffle the elements of p2 in the same way\n",
    "        parent_two.genes[points[1]], parent_two.genes[points[3]] = parent_two.genes[points[3]], parent_two.genes[points[1]]\n",
    "\n",
    "        # performing one point crossover\n",
    "        print(\"\\nParents after performing shuffles:\")\n",
    "        child_one, child_two = one_point_crossover(parent_one,parent_two, 0,1) \n",
    "\n",
    "        # unshuffling\n",
    "        child_one.genes[points[0]], child_one.genes[points[2]] = child_one.genes[points[2]], child_one.genes[points[0]] #reshuffle the elements of c1\n",
    "        child_one.genes[points[1]], child_one.genes[points[3]] = child_one.genes[points[3]], child_one.genes[points[1]]\n",
    "\n",
    "        child_two.genes[points[0]], child_two.genes[points[2]] = child_two.genes[points[2]], child_two.genes[points[0]] #reshuffle the elements of c2 in the same way\n",
    "        child_two.genes[points[1]], child_two.genes[points[3]] = child_two.genes[points[3]], child_two.genes[points[1]]\n",
    "        \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 1, 1, 0, 1, 0, 0, 1, 0])\n",
    "chrom1 = np.array([1, 0, 0, 0, 1, 0, 1, 1, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = shuffle_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='reduce-surrogate'>Reduce Surrogate Crossover (RSX)</a>\n",
    "\n",
    "This crossover happens only if the selected parents are different in more than one gene. This means that if the parents are similar or are different in only one gene, this crossover won't be operated on the parents. If the value of more than one gene is different, the positions where the genes of the parents have different values are selected and one of these positions is chosen randomly and one-point crossover is operated on parents in the chosen position."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/reduced_surrogate_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The chosen point for one_point_crossover: 1\n",
      "\n",
      "Parents in one_point_crossover:\n",
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 1 1 0 1 0 0 1 0]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 1 1 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 1 1 0]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[1 1 1 0 1 0 0 1 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def reduced_surragate_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Reduced surragate crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "   \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "    dif = 0\n",
    "\n",
    "    for i in range(chrom_length):\n",
    "        if parent_one.genes[i] != parent_two.genes[i]:\n",
    "            dif+=1\n",
    "  \n",
    "    # if pc is greater than random number and more that one gene is different\n",
    "    if np.random.rand() < pc and dif>1:  \n",
    "        # choose number of points to be nominated for crossover points\n",
    "        point_nums = np.random.randint(2, abs(chrom_length)) \n",
    "        # choose several positions for crossover\n",
    "        points = sorted_random_numbers(point_nums, chrom_length)\n",
    "        # choose the final point for 1PX \n",
    "        pos = np.random.randint(0, point_nums) \n",
    "        \n",
    "        print(\"The chosen point for one_point_crossover:\" , points[pos])\n",
    "        print(\"\\nParents in one_point_crossover:\")\n",
    "        child_one, child_two = one_point_crossover(parent_one, parent_two, points[pos], 1)        \n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 1, 1, 0, 1, 0, 0, 1, 0])\n",
    "chrom1 = np.array([1, 0, 0, 0, 1, 0, 1, 1, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = reduced_surragate_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='average'>Average Crossover (AX)</a>\n",
    "\n",
    "In average crossover, the genes of the offspring in each position are calculated by taking the arithmetic mean of both of its parents in that position. For integer representation, the floor of the arithmetic mean is considered for the value of the gene in the child.\n",
    "\n",
    "`child_one[i] = floor((parent_one[i] + parent_two[i]) / 2)`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/average_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[5 3 3 2 3 9 7 6 5 4]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[5 4 7 6 5 2 6 1 3 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[5 3 5 4 4 5 6 3 4 2]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def average_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Average crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")    \n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    # child_two = chromosome(genes= np.array([-1]*chrom_length),id=1,fitness = 125.2)\n",
    "  \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        for i in range(chrom_length):\n",
    "            child_one.genes[i] = int((parent_one.genes[i] + parent_two.genes[i])/2)\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "\n",
    "    return child_one\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([5, 3, 3, 2, 3, 9, 7, 6, 5, 4])\n",
    "chrom1 = np.array([5, 4, 7, 6, 5, 2, 6, 1, 3, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = average_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(CROSS)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='discrete'>Discrete Crossover (DC)</a>\n",
    "\n",
    "In discrete crossover, a uniform random real number is produced for each gene of the parent's chromosome. If this number is less than 0.5, then the gene of the offspring in that position will be filled by parent one. Otherwise, the gene of the offspring in that position is filled by parent two. The main difference between discrete crossover and uniform crossover is that in discrete crossover we don't calculate a mask vector for the crossover to take place and also one offspring is produced instead of two."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/discrete_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[1 1 1 0 1 0 0 1 0 1]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 1 1 0 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 0 1 0 1 0 1 1 0 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy \n",
    "\n",
    "def discrete_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Discrete crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")    \n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)  # child\n",
    "  \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        for i in range(chrom_length):\n",
    "            w = np.random.uniform(low=0, high=1)\n",
    "            if w<0.5:\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "            else:\n",
    "                child_one.genes[i] = parent_two.genes[i]\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "    \n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "\n",
    "    return child_one\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 1, 1, 0, 1, 0, 0, 1, 0, 1])\n",
    "chrom1 = np.array([1, 0, 0, 0, 1, 0, 1, 1, 0, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = discrete_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(CROSS)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='flat'>Flat Crossover (FC)</a>\n",
    "In flat crossover, for each gene of the child, we produce a uniform random real number w between the values of the parents and we transfer this number to the child. In other words, if the parent-one is the first parent and the parent-two is the second parent, then the value for the gene `i` of the child-one is calculated as follows:\n",
    "\n",
    "`child_one[i] ~ (minimum(parent_one[i], parent_two[i]), maximum(parent_one[i], parent_two[i]))`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/flat_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.121 0.152 0.231 0.143 0.732 0.315 0.434 0.633]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[0.765 0.168 0.914 0.435 0.893 0.332 0.981 0.194]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[0.167 0.164 0.604 0.241 0.755 0.332 0.677 0.612]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def flat_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Flat crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")    \n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0.0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "  \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        for i in range(chrom_length):\n",
    "            w = np.random.uniform(low= np.minimum(parent_one.genes[i], parent_two.genes[i]), high= np.maximum(parent_one.genes[i], parent_two.genes[i]))\n",
    "            child_one.genes[i] = round(w, 3)\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "\n",
    "    return child_one\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0.121, 0.152, 0.231, 0.143, 0.732, 0.315, 0.434, 0.633])\n",
    "chrom1 = np.array([0.765, 0.168, 0.914, 0.435, 0.893, 0.332, 0.981, 0.194])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = flat_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(CROSS)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='multivariate'>Multivariate Crossover</a>\n",
    "\n",
    "In multivariate crossover, we divide the parents into several substrings. For each substring, we produce a uniform random real number and we check whether that number is less than PC (crossover probability). If the random number produced is less than PC, then we operate one-point crossover for that substring and we transfer the resulted genes to the children. Otherwise, the genes from the parents in that substring are directly transferred from the parents to the offspring."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/multivariate_crossover.png\"  alt='missing' width=\"500\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 1 1 0 1 0 0 1 0]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 1 1 0]\n",
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=125.2, \n",
      "Genes=\n",
      "[0, 1, 0]\n",
      "ID=#1, Fitness=125.2, \n",
      "Genes=\n",
      "[1, 1, 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 1 1 0 1 0 0 1 0]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 1 1 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy \n",
    "\n",
    "def multivariate_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Multivariate crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")    \n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "    random_nums = []\n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        points = sorted_random_numbers(2,chrom_length) #selecting two points\n",
    "        for i in range(3): # selecting three random numbers for each substring\n",
    "            random_nums.append(np.random.rand())\n",
    "\n",
    "    #************************************************************************ part1\n",
    "        if random_nums[0]<0.5:\n",
    "            for i in range(points[0]):\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "        else:\n",
    "            parent1_sub = Chromosome(genes= [-1]*points[0],id_=0,fitness = 125.2)\n",
    "            parent2_sub = Chromosome(genes= [-1]*points[0],id_=1,fitness = 125.2)\n",
    "\n",
    "            for i in range(points[0]):\n",
    "                parent1_sub.genes[i] = parent_one.genes[i]\n",
    "                parent2_sub.genes[i] = parent_two.genes[i]\n",
    "\n",
    "\n",
    "            parent1_sub, parent2_sub = one_point_crossover(parent1_sub,parent2_sub,0,1) #perform 1PX\n",
    "\n",
    "            for i in range(points[0]): #transfer the genes to children\n",
    "                child_one.genes[i] = parent1_sub.genes[i]\n",
    "                child_two.genes[i] = parent2_sub.genes[i]\n",
    "\n",
    "    #************************************************************************part2\n",
    "        if random_nums[1]<0.5:\n",
    "            for i in range(points[0],points[1]):\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "        else:\n",
    "            parent1_sub = Chromosome(genes= [-1]*(points[1]-points[0]),id_=0,fitness = 125.2)\n",
    "            parent2_sub = Chromosome(genes= [-1]*(points[1]-points[0]),id_=1,fitness = 125.2)\n",
    "            k = 0\n",
    "            for i in range(points[0],points[1]):\n",
    "                parent1_sub.genes[k] = parent_one.genes[i]\n",
    "                parent2_sub.genes[k] = parent_two.genes[i]\n",
    "                k+=1\n",
    "\n",
    "            parent1_sub, parent2_sub = one_point_crossover(parent1_sub,parent2_sub,0,1) #perform 1PX\n",
    "            k = 0\n",
    "            for i in range(points[0],points[1]): #transfer the genes to children\n",
    "                child_one.genes[i] = parent1_sub.genes[k]\n",
    "                child_two.genes[i] = parent2_sub.genes[k]\n",
    "                k+=1\n",
    "\n",
    "    #************************************************************************part3\n",
    "        if random_nums[2]<0.5:\n",
    "            for i in range(points[1],chrom_length):\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "        else:\n",
    "            parent1_sub = Chromosome(genes= [-1]*(chrom_length-points[1]),id_=0,fitness = 125.2)\n",
    "            parent2_sub = Chromosome(genes= [-1]*(chrom_length-points[1]),id_=1,fitness = 125.2)\n",
    "            k = 0\n",
    "            for i in range(points[1],chrom_length):\n",
    "                parent1_sub.genes[k] = parent_one.genes[i]\n",
    "                parent2_sub.genes[k] = parent_two.genes[i]\n",
    "                k+=1\n",
    "\n",
    "            parent1_sub, parent2_sub = one_point_crossover(parent1_sub,parent2_sub,0,1) #perform 1PX\n",
    "            k = 0\n",
    "            for i in range(points[1],chrom_length): #transfer the genes to children\n",
    "                child_one.genes[i] = parent1_sub.genes[k]\n",
    "                child_two.genes[i] = parent2_sub.genes[k]\n",
    "                k+=1     \n",
    "      \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 1, 1, 0, 1, 0, 0, 1, 0])\n",
    "chrom1 = np.array([1, 0, 0, 0, 1, 0, 1, 1, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = multivariate_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='count-preserving'>Count Preserving Crossover (CPC)</a>\n",
    "\n",
    "**Count-preserving crossover (CPC)** is a genetic algorithm operator that aims to preserve the number of chromosomes equal to 1 in each chromosome in the initial population. <span style=\"color:#55CABB\">This operator is used to create offspring for the new population. It is used for problems such as the **traveling salesman problem**, to find the shortest possible route over generations. CPC is used in single-objective numerical optimization problems.</span>\n",
    "\n",
    "The CPC operator selects two parents from the parent pool and creates two lists of differences between them. The operator then tries to preserve the same number of genes equal to 1 in each chromosome in the initial population. Here's an example:\n",
    "\n",
    "```\n",
    "Parent 1: 1 0 1 0 1 1 1 0\n",
    "Parent 2: 1 1 1 0 1 0 0 1\n",
    "up: 5, 6\n",
    "down: 1, 7 \n",
    "```\n",
    "\n",
    "where up is the positions where the gene of parent-one has a value of 1 and the gene of parent-two has a value of 0; down is the positions where parent-two is 1 and parent-one is 0. The operator then creates two offspring by swapping the values in the up and down lists between the two parents. The resulting offspring are:\n",
    "\n",
    "```\n",
    "Offspring 1: 1 1 1 0 1 0 0 1\n",
    "Offspring 2: 1 0 1 0 1 1 1 0\n",
    "```\n",
    "\n",
    "As you can see, the offspring 1 is the same as parent 2, and offspring 2 is the same as parent 1; so we shuffle these values a little bit to avoid getting the same chromosomes over and over. So, the final offspring could be:\n",
    "\n",
    "```\n",
    "Offspring 1: 1 0 1 0 1 1 0 1\n",
    "Offspring 2: 1 0 1 0 1 0 1 1\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/cpx.png\"  alt='missing' width=\"500\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0 1 1 1 0 1 0 1 1 0]\n",
      "ID=#1, Fitness=6, \n",
      "Genes=\n",
      "[1 0 0 0 1 1 1 1 1 0]\n",
      "[1, 2, 3]\n",
      "[0, 4, 6]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[1 0 0 0 1 1 1 1 1 0]\n",
      "ID=#1, Fitness=6, \n",
      "Genes=\n",
      "[1 0 1 0 1 1 0 1 1 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy \n",
    "\n",
    "def count_preserving_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Count-preserving crossover on them.\n",
    "    In this crossover, the number of ones must be perserved in offsprings. We assume that \n",
    "    the number of ones are equal in both parents. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")    \n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "    up = [] #contains the positions of bits in the list which parent_one is one and parent_two is zero\n",
    "    down = [] #contains the positions of bits in the list which parent_one is zero and parent_two is one\n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "\n",
    "        # copying parents to children\n",
    "        for i in range(chrom_length): \n",
    "            child_one.genes[i] = parent_one.genes[i]\n",
    "            child_two.genes[i] = parent_two.genes[i]\n",
    "\n",
    "        for i in range(chrom_length):\n",
    "            if parent_one.genes[i] != parent_two.genes[i]:\n",
    "                if parent_one.genes[i] == 1 and parent_two.genes[i] == 0:\n",
    "                    up.append(i)\n",
    "                else:\n",
    "                    down.append(i)\n",
    "                    \n",
    "        \n",
    "        positions = up + down\n",
    "        positions.sort()\n",
    "\n",
    "        print(up)\n",
    "        print(down)\n",
    "\n",
    "        for i in range(len(up)): #changing the elements in the positions\n",
    "            child_one.genes[up[i]], child_one.genes[down[i]] = child_one.genes[down[i]], child_one.genes[up[i]] \n",
    "            child_two.genes[up[i]], child_two.genes[down[i]] = child_two.genes[down[i]], child_two.genes[up[i]] \n",
    "\n",
    "        #==================================================\n",
    "        # shuffling the lists of genes in children \n",
    "        #==================================================\n",
    "        child_1 = child_one.genes\n",
    "        child_2 = child_two.genes\n",
    "        # Create a new list with the values at the specified positions\n",
    "        values1 = [child_1[i] for i in positions]\n",
    "        values2 = [child_2[i] for i in positions]\n",
    "\n",
    "        # shuffle the lists to avoid producing similar children \n",
    "        np.random.shuffle(values1)\n",
    "        np.random.shuffle(values2)\n",
    "\n",
    "        # Replace the values in the original list with the shuffled values\n",
    "        for i, j in enumerate(positions):\n",
    "            child_1[j] = values1[i]\n",
    "            child_2[j] = values2[i]\n",
    "\n",
    "        child_one.genes = child_1\n",
    "        child_two.genes = child_2\n",
    "      \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 1, 1, 0, 1, 0, 1, 1, 0]) # same number of 1s\n",
    "chrom1 = np.array([1, 0, 0, 0, 1, 1, 1, 1, 1, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = count_preserving_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='modified-order'>Modified Order Crossover (MOC)</a>\n",
    "\n",
    "In modified order crossover, we choose a random position and then we find the left side of parent-two in parent-one and transfer these genes from parent-one to child-one. Then we loop over the right side of the parent-two and fill the empty positions of child-one from left to right in the same order. We do the same thing for child-two by finding the elements of the left side of parent-one in parent-two and transferring them; and then filling the empty spots with the elements on the right side of parent-one."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/moc.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 6 9 8 5 7]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[2 1 9 8 0 5 6 3 7 4]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[9 1 2 8 0 5 6 3 7 4]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[2 1 3 4 0 6 9 8 5 7]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy \n",
    "\n",
    "def modified_order_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Modified order crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")    \n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        hold1 = [] #for holding the elements in the one half of the parent_two\n",
    "        hold2 = [] #for holding the elements in the one half of the parent_one\n",
    "\n",
    "        pos = np.random.randint(2, chrom_length)\n",
    "        # pos = 4\n",
    "        for i in range(pos):\n",
    "            hold1.append(parent_two.genes[i]) #first half of the parent_two\n",
    "            hold2.append(parent_one.genes[i]) #first half of the parent_one\n",
    "        for i in range(chrom_length):\n",
    "            if parent_one.genes[i] in hold1:\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "            if parent_two.genes[i] in hold2:\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "\n",
    "        hold1 = []\n",
    "        hold2 = []\n",
    "        for i in range(pos,chrom_length):\n",
    "            hold1.append(parent_two.genes[i]) #second half of the parent_two\n",
    "            hold2.append(parent_one.genes[i]) #second half of the parent_one\n",
    "\n",
    "        k = 0\n",
    "        m = 0\n",
    "        for i in range(chrom_length):\n",
    "            if child_one.genes[i] == -1:\n",
    "                child_one.genes[i] = hold1[k]\n",
    "                k+=1\n",
    "            if child_two.genes[i] == -1:\n",
    "                child_two.genes[i] = hold2[m]\n",
    "                m+=1\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 2, 3, 4, 6, 9, 8, 5, 7])\n",
    "chrom1 = np.array([2, 1, 9, 8, 0, 5, 6, 3, 7, 4])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = modified_order_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='random-respectful'>Random Respectful Crossover (RRC)</a>\n",
    "\n",
    "In random respectful crossover, we use a list named `similarity vector` to produce the offspring. To calculate `similarity vector`, we compare the genes of the two parents in each position. If the value of the genes in that position is the same, the value of the similarity vector for that position becomes the value of the genes in that position. Otherwise, the value of the similarity vector for that position becomes -1. To calculate the values of the genes for the offspring we do the following.\n",
    "\n",
    "<ul>\n",
    "<li>If the value of the similarity vector in position i is 0, the value of the genes in that position for both offspring becomes 0.</li>\n",
    "<li>If the value of the similarity vector in position i is 1, the value of the genes in that position for both offspring becomes 1.</li>\n",
    "<li>If the value of the similarity vector in position i is -1, we produce a uniform random real number `w` between 0 and 1 to determine what the value of the offspring in that position would be.</li>\n",
    "    <ul type=\"circle\">\n",
    "        <li>If w < 0.5, the value of the gene in that position becomes 1;</li>\n",
    "        <li>otherwise, the value of the gene in that position becomes 0. </li>\n",
    "    </ul>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/random_respectful_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 1 1 0 1 0 0 1 0]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 1 1 0]\n",
      "\n",
      "Similarity-vector: [1, -1, -1, 0, 1, 0, -1, 1, 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 1 0 0 1 0 0 1 0]\n",
      "ID=#1, Fitness=3, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 0 1 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def calculate_similarity_vector(chrom1, chrom2):\n",
    "    \"\"\"\n",
    "    To calculate the similarity-vector for two individuals.\n",
    "    chrom1: First individual\n",
    "    chrom2: Second individual\n",
    "    \"\"\"  \n",
    "    chrom_length = Chromosome.get_chrom_length(chrom1)\n",
    "    similarity_vector = [-1]*chrom_length\n",
    "    for i in range(chrom_length):\n",
    "        if chrom1.genes[i] == chrom2.genes[i]:\n",
    "            if parent_one.genes[i] == 0:\n",
    "                similarity_vector[i] = 0\n",
    "            else:\n",
    "                similarity_vector[i] = 1\n",
    "    return similarity_vector\n",
    "\n",
    "\n",
    "def random_respectful_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Random respectful crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        \n",
    "        similarity_vector = calculate_similarity_vector(parent_one, parent_two)\n",
    "        print(\"\\nSimilarity-vector:\", similarity_vector)\n",
    "\n",
    "        for i in range(chrom_length):\n",
    "            if similarity_vector[i] == 1:\n",
    "                child_one.genes[i] = 1\n",
    "                child_two.genes[i] = 1\n",
    "            elif similarity_vector[i] == 0:\n",
    "                child_one.genes[i] = 0\n",
    "                child_two.genes[i] = 0\n",
    "            else:  #the similarity_vector[i] is -1\n",
    "                w = np.random.uniform(low=0, high=1) #for the first child\n",
    "                if w < 0.5:\n",
    "                    child_one.genes[i] = 1\n",
    "                else:\n",
    "                    child_one.genes[i] = 0\n",
    "\n",
    "                w = np.random.uniform(low=0, high=1) #for the second child\n",
    "                if w < 0.5:\n",
    "                    child_two.genes[i] = 1\n",
    "                else:\n",
    "                    child_two.genes[i] = 0\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 1, 1, 0, 1, 0, 0, 1, 0])\n",
    "chrom1 = np.array([1, 0, 0, 0, 1, 0, 1, 1, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = random_respectful_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='masked'>Masked Crossover</a>\n",
    "\n",
    "In masked crossover, we first copy the value of the genes from parent-one to child-one and from parent-two to child-two. In this crossover, we assume that each individual in the population has a `mask vector`. The values of the mask vector are calculated based on the individual's fitness. If the value of a mask vector in a position is 1 and the value of another mask vector in that position is 0, we say that the first mask vector is dominant in that position. If we consider `mask_vector1` as the mask vector of parent-one, and `mask_vector2` as the mask vector of parent-two, the offspring will be calculated as the following.\n",
    "> 1. If <span style=\"color:#EE7766\">**mask_vector2(i) = 1** and **mask_vector1(i) = 0**</span>, then **child_one(i) = parent_two(i)**.\n",
    "> 2. If <span style=\"color:#6677EE\">**mask_vector2(i) = 0** and **mask_vector1(i) = 1**</span>, then **child_two(i) = parent_one(i)**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/masked_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 1 1 0 1 0 0 1 0]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 1 1 0]\n",
      "\n",
      "mask vector for parent_one:  [1 0 1 0 1 1 0 1 0]\n",
      "mask vector for parent_two:  [1 1 1 0 1 0 1 1 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 0 1 0 1 0 1 1 0]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 1 1 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def masked_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Masked crossover on them.\n",
    "    In this crossover, each of the parents have a masked vector, and this vector for an individual \n",
    "    is calculated based on the fitness of that individual within the population.\n",
    "    In this method, we used vectors with random values for this purpose.    \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([0]*chrom_length),id_=1,fitness = 125.2)\n",
    "    \n",
    "    mask_one = np.random.choice([0, 1], size=chrom_length, p=[0.5, 0.5]) #mask vector for parent_one\n",
    "    mask_two = np.random.choice([0, 1], size=chrom_length, p=[0.5, 0.5]) #mask vector for parent_two\n",
    "    print(\"\\nmask vector for parent_one: \", mask_one)\n",
    "    print(\"mask vector for parent_two: \", mask_two)\n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        for i in range(chrom_length): #Copying parents into children \n",
    "            child_one.genes[i] = parent_one.genes[i]\n",
    "            child_two.genes[i] = parent_two.genes[i]\n",
    "\n",
    "        for i in range(chrom_length):\n",
    "            if mask_one[i] == 0 and mask_two[i] == 1: #The parent_two is dominant\n",
    "                child_one.genes[i] = parent_two.genes[i]\n",
    "            if mask_one[i] == 1 and mask_two[i] == 0: #The parent_one is dominant\n",
    "                child_two.genes[i] = parent_one.genes[i]\n",
    "            \n",
    "        \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 1, 1, 0, 1, 0, 0, 1, 0])\n",
    "chrom1 = np.array([1, 0, 0, 0, 1, 0, 1, 1, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = masked_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='1-bit-adaptive'>1-bit Adaptive Crossover (1BX)</a>\n",
    "\n",
    "In 1-bit adaptive crossover, we look at the value of the last gene in both of the parents. \n",
    "\n",
    "**(1)** If <span style=\"color:#EE7766\">the last two bits in the parents were the same and equal to 0</span>, then we perform **uniform crossover** on the parents.<br>\n",
    "**(2)** If <span style=\"color:#77EEAA\">the last two bits in the parents were the same and equal to 1</span>, then we perform **two-point crossover** on the parents.<br>\n",
    "**(3)** If <span style=\"color:#6767EE\">the last two bits in the parents were different</span>, we create a uniform random real number w to choose which crossover to perform. If w is less than 0.5, then we perform **uniform crossover**; otherwise, we perform **two-point crossover**.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/1bx.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 1 1 0 1 0 0 1 0]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 1 1 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=3, \n",
      "Genes=\n",
      "[1 0 0 0 1 0 0 1 0]\n",
      "ID=#1, Fitness=6, \n",
      "Genes=\n",
      "[1 1 1 0 1 0 1 1 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def claculate_mask(chrom_size, px):\n",
    "    \"\"\"\n",
    "    This method, performs the mask calculation for the Uniform crossover.\n",
    "    chrom_size: The mask has the same size of the chromosome.\n",
    "    px: A control parameter for the number of ones in the mask vector.\n",
    "    \"\"\"  \n",
    "    mask = [0]*chrom_size\n",
    "    for i in range(chrom_size):\n",
    "        prob = np.random.rand()\n",
    "        if prob <= px:\n",
    "            mask[i]=1\n",
    "    return mask\n",
    "\n",
    "def uniform_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Uniform crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([0]*chrom_length),id_=1,fitness = 125.2)\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        mask = claculate_mask(chrom_length,np.random.uniform(low=0.2, high=0.85))\n",
    "        for i in range(chrom_length):\n",
    "            if mask[i]==1:\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "            else:\n",
    "                child_one.genes[i] = parent_two.genes[i]\n",
    "                child_two.genes[i] = parent_one.genes[i]\n",
    "           \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "def one_bit_adaptive_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs 1_Bit adaptive crossover on them. \n",
    "    The last gene of two parents are the same and both zero ==> Uniform crossover\n",
    "    The last gene of two parents are the same and both one ==> Multi-point crossover  \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "    \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        \n",
    "        if parent_one.genes[-1] == 0 and parent_two.genes[-1] == 0:\n",
    "            child_one, child_two = uniform_crossover(parent_one,parent_two,1)\n",
    "        elif parent_one.genes[-1] == 1 and parent_two.genes[-1] == 1:\n",
    "            child_one, child_two = multi_point_crossover(parent_one,parent_two,1) \n",
    "        else:\n",
    "            w = np.random.uniform(low=0, high=1)\n",
    "            if w < 0.5:\n",
    "                child_one, child_two = uniform_crossover(parent_one,parent_two,1)\n",
    "            else:\n",
    "                child_one, child_two = multi_point_crossover(parent_one,parent_two,1) \n",
    "                \n",
    "        \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "    \n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 1, 1, 0, 1, 0, 0, 1, 0])\n",
    "chrom1 = np.array([1, 0, 0, 0, 1, 0, 1, 1, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "\n",
    "CROSS = one_bit_adaptive_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='modified-partially-mapped'>Modified Partially Mapped Crossover (MPMX) </a>\n",
    "\n",
    "In modified partially mapped crossover, we first produce two randomly selected positions. We transfer every gene between these two points directly from parent-one to child-one and from parent-two to child-two. Then, we do the following:\n",
    "\n",
    "(1). <span style=\"color:#998800\">**For child-one**</span>, we fill the positions that can be filled (the elements that don't already exist in child-one) by directly transferring the genes in these positions from parent-two to child-one. After that, we shuffle the remaining elements and fill the empty positions in child-one with them. \n",
    "\n",
    "(2). <span style=\"color:#008877\"> **For child-two**</span>, we fill the positions that can be filled by directly transferring the genes in those positions from parent-one to child-two. Then, we shuffle the remaining elements and fill the empty positions in child-two with them."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/modified_partially_mapped_crossover.png\"  alt='missing' width=\"700\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 8 4 5 6 7 1 2 3 9]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[6 7 1 2 4 8 3 5 9 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[2 8 4 5 6 7 3 1 9 0]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[0 7 1 2 4 8 5 6 3 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def modified_partially_mapped_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Modified partially mapped crossover on them.  \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "    points = sorted_random_numbers(2,chrom_length)\n",
    "    # points[0] = 2\n",
    "    # points[1] = 6\n",
    "    seen_list1 = [] #Elements that are transfered to child_one\n",
    "    seen_list2 = [] #Elements that are transfered to child_two\n",
    "    not_seen_list1 = [] #Elements that are not transfered to child_one\n",
    "    not_seen_list2 = [] #Elements that are not transfered to child_two\n",
    "    \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        for i in range(points[0], points[1]):\n",
    "            child_one.genes[i] = parent_one.genes[i]\n",
    "            seen_list1.append(parent_one.genes[i])\n",
    "            child_two.genes[i] = parent_two.genes[i]\n",
    "            seen_list2.append(parent_two.genes[i])\n",
    "\n",
    "        for i in range(chrom_length):\n",
    "            if i < points[0] or i >= points[1]:\n",
    "                if parent_two.genes[i] not in seen_list1:\n",
    "                    child_one.genes[i] = parent_two.genes[i]\n",
    "                    seen_list1.append(parent_two.genes[i])\n",
    "                if parent_one.genes[i] not in seen_list2:\n",
    "                    child_two.genes[i] = parent_one.genes[i]\n",
    "                    seen_list2.append(parent_one.genes[i])\n",
    "        \n",
    "        for i in range(chrom_length):\n",
    "            if parent_one.genes[i] not in seen_list1:\n",
    "                not_seen_list1.append(parent_one.genes[i])\n",
    "            if parent_one.genes[i] not in seen_list2:\n",
    "                not_seen_list2.append(parent_one.genes[i])\n",
    "\n",
    "        np.random.shuffle(not_seen_list1) #shuffling the elements that are not present in children \n",
    "        np.random.shuffle(not_seen_list2)\n",
    "        k = 0\n",
    "        m = 0\n",
    "        for i in range(chrom_length):\n",
    "            if child_one.genes[i] == -1:\n",
    "                child_one.genes[i] = not_seen_list1[k]\n",
    "                k+=1\n",
    "            if child_two.genes[i] == -1:\n",
    "                child_two.genes[i] = not_seen_list2[m]\n",
    "                m+=1                      \n",
    "        \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 8, 4, 5, 6, 7, 1, 2, 3, 9])\n",
    "chrom1 = np.array([6, 7, 1, 2, 4, 8, 3, 5, 9, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = modified_partially_mapped_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='order-based'>Order-Based Crossover (OBX)</a>\n",
    "\n",
    "In order-based crossover, we choose **three random genes** from parent-two and we find their place in parent-one. We keep these places in child-one empty and fill every other position in child-one by transferring the genes from parent-one directly to them. Then, we loop over parent-two and transfer to child-one the genes that are not already transferred to it. \n",
    "\n",
    "For child two, we already saved the chosen places in parent-one, so we keep the same places empty in child-two and fill every other position in child-two by transferring the genes from parent-two directly to them. Then, we loop over parent-one and transfer to child-two the genes that are not already transferred to it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/order_based_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 8 4 5 6 7 1 2 3 9]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[6 7 1 2 4 8 3 5 9 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 8 2 5 6 7 1 4 3 9]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[6 7 5 2 4 8 3 1 9 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def ordered_based_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Ordered based crossover on them.  \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "\n",
    "    not_seen_list1 = [] #Elements that are not transfered to child_one\n",
    "    not_seen_list2 = [] #Elements that are not transfered to child_two\n",
    "    \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        points = sorted_random_numbers(3,chrom_length) \n",
    "        # points[0] = 3\n",
    "        # points[1] = 4\n",
    "        # points[2] = 5      \n",
    "        for i in range(len(points)):\n",
    "            not_seen_list1.append(parent_one.genes[points[i]])\n",
    "            not_seen_list2.append(parent_two.genes[points[i]])\n",
    "        \n",
    "        for i in range(chrom_length): #transfering genes of parents to children except chosen positions\n",
    "            if i not in points:\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "\n",
    "        k = 0\n",
    "        for i in range(chrom_length):\n",
    "            if parent_two.genes[i] in not_seen_list1: #elements in parent_two that are not in child_one\n",
    "                child_one.genes[points[k]] = parent_two.genes[i]\n",
    "                k+=1\n",
    "        k = 0\n",
    "        for i in range(chrom_length):\n",
    "            if parent_one.genes[i] in not_seen_list2: #elements in parent_one that are not in child_two\n",
    "                child_two.genes[points[k]] = parent_one.genes[i]\n",
    "                k+=1\n",
    "\n",
    "                          \n",
    "        \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 8, 4, 5, 6, 7, 1, 2, 3, 9])\n",
    "chrom1 = np.array([6, 7, 1, 2, 4, 8, 3, 5, 9, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = ordered_based_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='position-based'>Position-Based Crossover (POS)</a>\n",
    "\n",
    "In position-based crossover, we choose several random genes from parent-two and directly transfer them to child-one. Then, we fill the empty positions in child one, by looping over parent-one and transferring all the genes that are not already transferred to child-one from parent-two. \n",
    "\n",
    "For child-two, we transfer the genes in the positions that were randomly chosen earlier, from parent-one directly to child-two. Then we loop over the parent-two and transfer all the genes that are not already transferred to child-two from parent-one."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/position_based_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 8 4 5 6 7 1 2 3 9]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[6 7 1 2 4 8 3 5 9 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[8 7 5 6 4 1 2 3 9 0]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[7 8 1 2 6 4 3 5 0 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def position_based_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Position based crossover on them.  \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "\n",
    "    seen_list1 = [] #Elements that are transfered to child_one\n",
    "    seen_list2 = [] #Elements that are transfered to child_two\n",
    "    \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        \"points in the begining, is the positions that are to be transfered from parent_two to child_one.\"\n",
    "        \"Also the positions that are to be transfered from parent_one to child_two.\"\n",
    "        points = sorted_random_numbers(3,chrom_length)  \n",
    "        # points[0] = 1\n",
    "        # points[1] = 2\n",
    "        # points[2] = 5      \n",
    "        for i in range(len(points)):\n",
    "            seen_list1.append(parent_two.genes[points[i]])\n",
    "            seen_list2.append(parent_one.genes[points[i]])\n",
    "        \n",
    "        for i in range(chrom_length): #transfering genes of parents to children in the chosen positions\n",
    "            if i in points:\n",
    "                child_one.genes[i] = parent_two.genes[i]\n",
    "                child_two.genes[i] = parent_one.genes[i]\n",
    "\n",
    "        points = [] #positions in the children that are not already filled\n",
    "        for i in range(chrom_length):\n",
    "            if child_one.genes[i] == -1:\n",
    "                points.append(i)\n",
    "        \n",
    "        k = 0\n",
    "        for i in range(chrom_length):\n",
    "            if parent_one.genes[i] not in seen_list1: #elements in parent_one that are not in child_one\n",
    "                child_one.genes[points[k]] = parent_one.genes[i]\n",
    "                k+=1\n",
    "        \n",
    "        k = 0\n",
    "        for i in range(chrom_length):\n",
    "            if parent_two.genes[i] not in seen_list2: #elements in parent_two that are not in child_two\n",
    "                child_two.genes[points[k]] = parent_two.genes[i]\n",
    "                k+=1\n",
    "                           \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 8, 4, 5, 6, 7, 1, 2, 3, 9])\n",
    "chrom1 = np.array([6, 7, 1, 2, 4, 8, 3, 5, 9, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "\n",
    "CROSS = position_based_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='voting-recombination'>Voting Recombination Crossover (VR)</a>\n",
    "\n",
    "In voting recombination crossover, we have **several parents** and a positive integer as a `threshold` which is less than or equal to the number of parents. If the value of a gene in a position has been the same in a number of parents equal to or greater than the threshold, then the value of the gene in that position for the offspring will become the same as that value. The remaining elements are shuffled and transferred to the offspring. <br>\n",
    "In the following example, we have four parents and the threshold is 3."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/voting_recombination_crossover.png\"  alt='missing' width=\"600\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 4 3 5 2 6 0]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[1 2 0 3 5 6 4]\n",
      "ID=#2, Fitness=5, \n",
      "Genes=\n",
      "[3 2 1 5 4 0 6]\n",
      "ID=#3, Fitness=5, \n",
      "Genes=\n",
      "[1 2 3 4 0 6 5]\n",
      "ID=#4, Fitness=5, \n",
      "Genes=\n",
      "[0 4 2 1 6 5 3]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 2 5 4 3 6 0]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[1 2 4 5 3 6 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "from collections import Counter\n",
    "\n",
    "def most_frequent(List):\n",
    "    \"\"\"\n",
    "    Returns the element with most occurrence in a list\n",
    "    \"\"\" \n",
    "    occurence_count = Counter(List) \n",
    "    return occurence_count.most_common(1)[0][0] \n",
    "\n",
    "\n",
    "def voting_recombination_crossover(parents, pc):\n",
    "    \"\"\"\n",
    "    This function takes several parents, and performs Voting recombination crossover on them. \n",
    "    In this method, threshold is the number of repeatition for an element in one position, for \n",
    "    the child would be the same element, in that position. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    chrom_length = Chromosome.get_chrom_length(parents[0])\n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    for i in range(len(parents)):\n",
    "        Chromosome.describe(parents[i])\n",
    "\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "    threshold = 3\n",
    "    seen_list = [] #Elements that are transfered to child\n",
    "    not_seen_list = [] #Elements that are not transfered to child\n",
    "    \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        for i in range(chrom_length):\n",
    "            column = []\n",
    "            for j in range(len(parents)):\n",
    "                column.append(parents[j].genes[i])\n",
    "\n",
    "            vote = most_frequent(column)\n",
    "            votes_num = column.count(vote)\n",
    "            if votes_num >= threshold:\n",
    "                child_one.genes[i] = vote\n",
    "                child_two.genes[i] = vote\n",
    "                seen_list.append(vote)\n",
    "        \n",
    "        for i in range(chrom_length):\n",
    "            if parents[0].genes[i] not in seen_list:\n",
    "                not_seen_list.append(parents[0].genes[i])\n",
    "\n",
    "        np.random.shuffle(not_seen_list)     \n",
    "        k = 0\n",
    "        for i in range(chrom_length): #filling empty elements of first child\n",
    "            if child_one.genes[i] == -1:\n",
    "                child_one.genes[i] = not_seen_list[k]\n",
    "                k+=1\n",
    "\n",
    "        np.random.shuffle(not_seen_list)     \n",
    "        k = 0\n",
    "        for i in range(chrom_length): #filling empty elements of second child\n",
    "            if child_two.genes[i] == -1:\n",
    "                child_two.genes[i] = not_seen_list[k]\n",
    "                k+=1\n",
    "     \n",
    "    else:  # if pc is less than random number then the children will be one of the parents involved\n",
    "        child_one = deepcopy(parents[np.random.randint(0, len(parents))])\n",
    "        child_two = deepcopy(parents[np.random.randint(0, len(parents))])\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "# parents\n",
    "parents = []\n",
    "genes0 = np.array([1, 4, 3, 5, 2, 6, 0])\n",
    "genes1 = np.array([1, 2, 0, 3, 5, 6, 4])\n",
    "genes2 = np.array([3, 2, 1, 5, 4, 0, 6])\n",
    "genes3 = np.array([1, 2, 3, 4, 0, 6, 5])\n",
    "genes4 = np.array([0, 4, 2, 1, 6, 5, 3])\n",
    "\n",
    "parents.append(Chromosome(genes= np.array(genes0), id_=0, fitness = fitness_function(genes0)))\n",
    "parents.append(Chromosome(genes= np.array(genes1), id_=1, fitness = fitness_function(genes1)))\n",
    "parents.append(Chromosome(genes= np.array(genes2), id_=2, fitness = fitness_function(genes2)))\n",
    "parents.append(Chromosome(genes= np.array(genes3), id_=3, fitness = fitness_function(genes3)))\n",
    "parents.append(Chromosome(genes= np.array(genes4), id_=4, fitness = fitness_function(genes4)))\n",
    "\n",
    "CROSS = voting_recombination_crossover(parents, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='maximal-preservation'>Maximal Preservation Crossover (MPX)</a>\n",
    "\n",
    "In maximal preservation crossover, we select **a randomly appropriately sized substring** from parent-one and transfer the genes in that substring directly from parent-one to child-one. The transferred elements are deleted from parent-two and the remaining elements are transferred from parent-two to child-one. \n",
    "\n",
    "For child-two, in the same way, we select a randomly appropriately sized substring from parent-two and transfer the genes in that substring directly from parent-two to child-one. The transferred elements are deleted from parent-one and the remaining elements are transferred from parent-one to child-two."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/mpx.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[1 4 3 2 6 7 5 9 8 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 4 3 2 6 5 9 7 8 0]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[0 1 3 2 6 7 4 5 8 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def maximal_preservation_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Maximal perservation crossover on them.  \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "\n",
    "    \n",
    "    seen_list1 = [] #Elements that are transfered to child_one\n",
    "    seen_list2 = [] #Elements that are transfered to child_two\n",
    "    \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        #********************************************************* Filling Child_one\n",
    "        points = sorted_random_numbers(2, chrom_length) #for choosing a substring\n",
    "        # points[0] = 1\n",
    "        # points[1] = 5\n",
    "        for i in range(points[0], points[1]): #transfering the substring from parent_one to child_one\n",
    "            child_one.genes[i] = parent_one.genes[i]\n",
    "            seen_list1.append(parent_one.genes[i])\n",
    "        \n",
    "        points = [] #finding empty spots in child_one\n",
    "        for i in range(chrom_length):\n",
    "            if child_one.genes[i] == -1:\n",
    "                points.append(i)\n",
    "        \n",
    "        k = 0\n",
    "        for i in range(chrom_length):\n",
    "            if parent_two.genes[i] not in seen_list1:\n",
    "                child_one.genes[points[k]] = parent_two.genes[i]\n",
    "                k+=1\n",
    "        #********************************************************* Filling Child_two\n",
    "        points = sorted_random_numbers(2, chrom_length) #for choosing a substring\n",
    "        # points[0] = 2\n",
    "        # points[1] = 6\n",
    "        for i in range(points[0], points[1]): #transfering the substring from parent_one to child_one\n",
    "            child_two.genes[i] = parent_two.genes[i]\n",
    "            seen_list2.append(parent_two.genes[i])\n",
    "        \n",
    "        points = [] #finding empty spots in child_two\n",
    "        for i in range(chrom_length):\n",
    "            if child_two.genes[i] == -1:\n",
    "                points.append(i)\n",
    "        \n",
    "        k = 0\n",
    "        for i in range(chrom_length):\n",
    "            if parent_one.genes[i] not in seen_list2:\n",
    "                child_two.genes[points[k]] = parent_one.genes[i]\n",
    "                k+=1\n",
    "                           \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "    \n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "chrom1 = np.array([1, 4, 3, 2, 6, 7, 5, 9, 8, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = maximal_preservation_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='position'>Position Crossover (PX)</a>\n",
    "\n",
    "In position crossover, several positions are randomly selected in parent-one and transferred to child-one. The transferred elements are deleted from parents-two and by a loop on the remaining elements of parents-two, the elements that are not already transferred to child-one are transferred to it.\n",
    "\n",
    "For child two, in the same way, several positions are randomly selected in parent-two and transferred to child-two. The transferred elements are deleted from parent-one and by a loop on the remaining elements of parents-one, the elements that are not already transferred to child-two are transferred to it. \n",
    "\n",
    "<span style=\"color:#EE9977\">The main difference between position crossover and maximal preservation crossover is that in the position crossover the randomly selected elements are not necessarily in a sequence.</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/position_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[1 4 3 2 6 7 5 9 8 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[4 1 3 2 6 5 8 7 0 9]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[0 1 3 2 4 7 5 9 6 8]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def position_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Position crossover on them.\n",
    "    In this method, we assumong the length of chromosomes are more than 5.\n",
    "    The main difference of this crossover with Maximal-preservation crossover, is that the points \n",
    "    that are being transfered directly to children, are not necessarily consequent.   \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "  \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "\n",
    "    \n",
    "    seen_list1 = [] #Elements that are transfered to child_one\n",
    "    seen_list2 = [] #Elements that are transfered to child_two\n",
    "    \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        #********************************************************* Filling Child_one\n",
    "        point_nums = np.random.randint(2,chrom_length-2) #number of points\n",
    "        points = sorted_random_numbers(point_nums, chrom_length) #for choosing several points\n",
    "\n",
    "        for i in range(chrom_length):\n",
    "            if i in points:\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "                seen_list1.append(parent_one.genes[i])\n",
    "        \n",
    "        points = [] #finding empty spots in child_one\n",
    "        for i in range(chrom_length):\n",
    "            if child_one.genes[i] == -1:\n",
    "                points.append(i)\n",
    "        k = 0\n",
    "        for i in range(chrom_length):\n",
    "            if parent_two.genes[i] not in seen_list1:\n",
    "                child_one.genes[points[k]] = parent_two.genes[i]\n",
    "                k+=1\n",
    "        #********************************************************* Filling Child_two\n",
    "        point_nums = np.random.randint(2,chrom_length-2) #number of points\n",
    "        points = sorted_random_numbers(point_nums, chrom_length) #for choosing several points\n",
    "\n",
    "        for i in range(chrom_length):\n",
    "            if i in points:\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "                seen_list2.append(parent_two.genes[i])\n",
    "        \n",
    "        points = [] #finding empty spots in child_two\n",
    "        for i in range(chrom_length):\n",
    "            if child_two.genes[i] == -1:\n",
    "                points.append(i)\n",
    "        k = 0\n",
    "        for i in range(chrom_length):\n",
    "            if parent_one.genes[i] not in seen_list2:\n",
    "                child_two.genes[points[k]] = parent_one.genes[i]\n",
    "                k+=1\n",
    "                           \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "chrom1 = np.array([1, 4, 3, 2, 6, 7, 5, 9, 8, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = position_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='homologous'>Homologous Crossover (HX)</a>\n",
    "\n",
    "In homologous crossover, we divide the parents into several substrings and only those substrings with enough similarity are allowed to be performed multi-point crossover on them. To calculate the similarity, we define a vector named `similarity vector` and a variable named `threshold`. The similarity vector is calculated by performing exclusive or (XOR) on the genes of the parents.\n",
    "\n",
    "A `1` value in the similarity vector means that the value of the genes in that place is different and the `0` value means that the value of the genes in that place is the same. If the number of 1s in a substring in the similarity vector is less than the threshold, it means that substring has enough similarity to be allowed for multi-point crossover."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/homologous_crossover.png\"  alt='missing' width=\"600\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0 1 0 0 0 1 0 1 1 1 1]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[1 0 1 1 0 1 0 1 0 0 0]\n",
      "[1, 3, 3]\n",
      "[False, True, True]\n",
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=125.2, \n",
      "Genes=\n",
      "[1, 0, 0, 0, 1, 0, 1, 1, 1, 1]\n",
      "ID=#1, Fitness=125.2, \n",
      "Genes=\n",
      "[0, 1, 1, 0, 1, 0, 1, 0, 0, 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 1 0 0 1 0 1 0 1 0]\n",
      "ID=#1, Fitness=6, \n",
      "Genes=\n",
      "[1 0 0 1 0 1 0 1 1 0 1]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def homologous_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Homologous crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "\n",
    "    child_one.genes = parent_one.genes #copying parents to their children\n",
    "    child_two.genes = parent_two.genes\n",
    "\n",
    "  \n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        point1 = np.random.randint(0, int(chrom_length/2) - 1)\n",
    "        point2 = np.random.randint(int(chrom_length/2) + 1, chrom_length)\n",
    "        sum1 = 0\n",
    "        sum2 = 0\n",
    "        sum3 = 0\n",
    "        sums = [] #the differences in each part of parents\n",
    "        yardstick = [] #parts of parents that the diversity is greater than threshold to participate in multi_point_crossover\n",
    "        threshold = 2\n",
    "        similarity_vector = []\n",
    "        for i in range(chrom_length):\n",
    "            similarity_vector.append(parent_one.genes[i] ^ parent_two.genes[i])\n",
    "        # print(similarity_vector)\n",
    "        for i in range(chrom_length):\n",
    "            if i < point1:\n",
    "                sum1 = sum1 + similarity_vector[i]\n",
    "            elif i >= point1 and i < point2:\n",
    "                sum2 = sum2 + similarity_vector[i]\n",
    "            else:\n",
    "                sum3 = sum3 + similarity_vector[i]\n",
    "\n",
    "        sums.append(sum1)\n",
    "        sums.append(sum2)\n",
    "        sums.append(sum3)\n",
    "        print(sums)\n",
    "        for i in range(len(sums)): #adding the parts with enough diversity\n",
    "            if sums[i] >= threshold:\n",
    "                yardstick.append(True)\n",
    "            else:\n",
    "                yardstick.append(False)\n",
    "\n",
    "        print(yardstick)\n",
    "        a = []\n",
    "        b = []\n",
    "        sub_parent_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "        sub_parent_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)        \n",
    "\n",
    "        #*************************************************************************** Eigth Different States\n",
    "        if yardstick[0] == True and yardstick[1] == True and yardstick[2] == True: #State 1\n",
    "            child_one, child_two = multi_point_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "        elif yardstick[0] == True and yardstick[1] == True and yardstick[2] == False: #State 2\n",
    "            for i in range(0, point2):\n",
    "                a.append(parent_one.genes[i])\n",
    "                b.append(parent_two.genes[i])\n",
    "\n",
    "            sub_parent_one.genes = a\n",
    "            sub_parent_two.genes = b\n",
    "\n",
    "            sub_parent_one, sub_parent_two = multi_point_crossover(sub_parent_one, sub_parent_two, 1)\n",
    "\n",
    "            k = 0\n",
    "            for i in range(0, point2):\n",
    "                child_one.genes[i] = sub_parent_one.genes[k]\n",
    "                child_two.genes[i] = sub_parent_two.genes[k]\n",
    "                k+=1\n",
    "            \n",
    "        elif yardstick[0] == True and yardstick[1] == False and yardstick[2] == True: #State 3\n",
    "            for i in range(0, point1):\n",
    "                a.append(parent_one.genes[i])\n",
    "                b.append(parent_two.genes[i])\n",
    "\n",
    "            for i in range(point2, chrom_length):\n",
    "                a.append(parent_one.genes[i])\n",
    "                b.append(parent_two.genes[i])\n",
    "\n",
    "            sub_parent_one.genes = a\n",
    "            sub_parent_two.genes = b\n",
    "\n",
    "            sub_parent_one, sub_parent_two = multi_point_crossover(sub_parent_one, sub_parent_two, 1)\n",
    "            \n",
    "            k = 0\n",
    "            for i in range(0, point1):\n",
    "                child_one.genes[i] = sub_parent_one.genes[k]\n",
    "                child_two.genes[i] = sub_parent_two.genes[k]\n",
    "                k+=1                  \n",
    "            for i in range(point2, chrom_length):\n",
    "                child_one.genes[i] = sub_parent_one.genes[k]\n",
    "                child_two.genes[i] = sub_parent_two.genes[k]\n",
    "                k+=1\n",
    "\n",
    "        elif yardstick[0] == False and yardstick[1] == True and yardstick[2] == True: #State 4\n",
    "            for i in range(point1, chrom_length):\n",
    "                a.append(parent_one.genes[i])\n",
    "                b.append(parent_two.genes[i])\n",
    "\n",
    "            sub_parent_one.genes = a\n",
    "            sub_parent_two.genes = b\n",
    "\n",
    "            sub_parent_one, sub_parent_two = multi_point_crossover(sub_parent_one, sub_parent_two, 1)\n",
    "\n",
    "            k = 0\n",
    "            for i in range(point1, chrom_length):\n",
    "                child_one.genes[i] = sub_parent_one.genes[k]\n",
    "                child_two.genes[i] = sub_parent_two.genes[k]\n",
    "                k+=1\n",
    "\n",
    "        elif yardstick[0] == False and yardstick[1] == False and yardstick[2] == True: #State 5\n",
    "            for i in range(point2, chrom_length):\n",
    "                a.append(parent_one.genes[i])\n",
    "                b.append(parent_two.genes[i])\n",
    "\n",
    "            sub_parent_one.genes = a\n",
    "            sub_parent_two.genes = b\n",
    "\n",
    "            sub_parent_one, sub_parent_two = multi_point_crossover(sub_parent_one, sub_parent_two, 1)\n",
    "\n",
    "            k = 0\n",
    "            for i in range(point2, chrom_length):\n",
    "                child_one.genes[i] = sub_parent_one.genes[k]\n",
    "                child_two.genes[i] = sub_parent_two.genes[k]\n",
    "                k+=1\n",
    "\n",
    "        elif yardstick[0] == False and yardstick[1] == True and yardstick[2] == False: #State 6\n",
    "            for i in range(point1, point2):\n",
    "                a.append(parent_one.genes[i])\n",
    "                b.append(parent_two.genes[i])\n",
    "\n",
    "            sub_parent_one.genes = a\n",
    "            sub_parent_two.genes = b\n",
    "\n",
    "            sub_parent_one, sub_parent_two = multi_point_crossover(sub_parent_one, sub_parent_two, 1)\n",
    "\n",
    "            k = 0\n",
    "            for i in range(point1, point2):\n",
    "                child_one.genes[i] = sub_parent_one.genes[k]\n",
    "                child_two.genes[i] = sub_parent_two.genes[k]\n",
    "                k+=1\n",
    "\n",
    "        elif yardstick[0] == True and yardstick[1] == False and yardstick[2] == False: #State 7\n",
    "            for i in range(0, point1):\n",
    "                a.append(parent_one.genes[i])\n",
    "                b.append(parent_two.genes[i])\n",
    "\n",
    "            sub_parent_one.genes = a\n",
    "            sub_parent_two.genes = b\n",
    "\n",
    "            sub_parent_one, sub_parent_two = multi_point_crossover(sub_parent_one, sub_parent_two, 1)\n",
    "\n",
    "            k = 0\n",
    "            for i in range(0, point1):\n",
    "                child_one.genes[i] = sub_parent_one.genes[k]\n",
    "                child_two.genes[i] = sub_parent_two.genes[k]\n",
    "                k+=1\n",
    "\n",
    "\n",
    "        #**************************************************************************************************\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1])\n",
    "chrom1 = np.array([1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = homologous_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='complete-subtour-exchange'>Complete Subtour Exchange Crossover (CSEX)</a>\n",
    "\n",
    "In complete subtour exchange crossover, equal or symmetric (reversed) substrings are found in the two parents and several offspring are produced based on different modes of these substrings. Two fittest of these offspring are returned as the offspring."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/complete_subtour_exchange_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[4 9 7 6 5 0 8 2 1 3]\n",
      "\n",
      "Children: [[0, 2, 1, 3, 4, 5, 6, 7, 8, 9], [0, 1, 2, 3, 4, 7, 6, 5, 8, 9], [4, 9, 7, 6, 5, 0, 8, 1, 2, 3], [4, 9, 5, 6, 7, 0, 8, 2, 1, 3], [0, 2, 1, 3, 4, 7, 6, 5, 8, 9], [4, 9, 5, 6, 7, 0, 8, 1, 2, 3]]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[4 9 5 6 7 0 8 2 1 3]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[0 2 1 3 4 7 6 5 8 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def complete_subtour_exchange_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Complete subtour exchange crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        list1 = [] #for saving the positions of the subset, in parent_one\n",
    "        list2 = [] #for saving the positions of the subset, in parent_two\n",
    "        total1 = [] #for saving all the positions of the subsets, in parent_one\n",
    "        total2 = [] #for saving all the positions of the subsets, in parent_two\n",
    "        for i in range(chrom_length):\n",
    "            gene_index = parent_two.genes.tolist().index(parent_one.genes[i])\n",
    "            # print(gene_index)\n",
    "            if gene_index + 1 < chrom_length and i + 1 < chrom_length:\n",
    "                if parent_two.genes[gene_index + 1] == parent_one.genes[i + 1]:\n",
    "                    k = i\n",
    "                    list1.append(k)\n",
    "                    list2.append(gene_index)\n",
    "                    for j in range(gene_index, chrom_length): #forward loop\n",
    "                        if parent_two.genes[j] == parent_one.genes[k]:\n",
    "                            k+=1\n",
    "                        else:\n",
    "                            break\n",
    "                    \n",
    "                    list1.append(k)\n",
    "                    list2.append(j + 1)\n",
    "                    total1.append(list1)\n",
    "                    total2.append(list2)\n",
    "                    list1 = []\n",
    "                    list2 = []\n",
    "        for i in range(chrom_length):\n",
    "            gene_index = parent_two.genes.tolist().index(parent_one.genes[i])\n",
    "            # print(gene_index)\n",
    "            list1 = []\n",
    "            list2 = []\n",
    "            if gene_index - 1 >= 0 and i + 1 < chrom_length:\n",
    "                if parent_two.genes[gene_index - 1] == parent_one.genes[i + 1]:\n",
    "                    k = i\n",
    "                    list1.append(k)\n",
    "                    list2.append(gene_index + 1)\n",
    "                    for j in range(gene_index, -1, -1): #backward loop\n",
    "                        if parent_two.genes[j] == parent_one.genes[k]:\n",
    "                            k+=1\n",
    "                        else:\n",
    "                            break\n",
    "\n",
    "                    list1.append(k)\n",
    "                    list2.append(j + 1)\n",
    "                    total1.append(list1)\n",
    "                    total2.append(list2)\n",
    "                    list1 = []\n",
    "                    list2 = []\n",
    "\n",
    "        for i in range(len(total2)): #correcting the indexes in total2\n",
    "            if total2[i][0] > total2[i][1]:\n",
    "                total2[i][0], total2[i][1] = total2[i][1], total2[i][0]\n",
    "        \n",
    "        #************************************************* Deleting redundant subsets\n",
    "        seen_list = []\n",
    "        indexes = []\n",
    "        for i in range(len(total1)): \n",
    "            if total1[i][1] not in seen_list:\n",
    "                seen_list.append(total1[i][1])\n",
    "            else:\n",
    "                indexes.append(i)\n",
    "\n",
    "        # print(indexes)\n",
    "        for i in range(len(indexes)):\n",
    "            total1.remove(total1[indexes[i]])\n",
    "            total2.remove(total2[indexes[i]])\n",
    "        #************************************************************ Children with one reversed subset\n",
    "        children = []\n",
    "        for i in range(len(total1)):\n",
    "            interval = total1[i]\n",
    "            child = parent_one.genes.tolist()\n",
    "            reversed_list = []\n",
    "            for j in range(interval[0], interval[1]):\n",
    "                reversed_list.append(child[j])\n",
    "            reversed_list.reverse()\n",
    "            k = 0\n",
    "            for j in range(interval[0], interval[1]):\n",
    "                child[j] = reversed_list[k]\n",
    "                k+=1\n",
    "            children.append(child)      \n",
    "\n",
    "        for i in range(len(total2)):\n",
    "            interval = total2[i]\n",
    "            child = parent_two.genes.tolist()\n",
    "            reversed_list = []\n",
    "            for j in range(interval[0], interval[1]):\n",
    "                reversed_list.append(child[j])\n",
    "            reversed_list.reverse()\n",
    "            k = 0\n",
    "            for j in range(interval[0], interval[1]):\n",
    "                child[j] = reversed_list[k]\n",
    "                k+=1\n",
    "            children.append(child) \n",
    "    #****************************************************************** Adding the other children   \n",
    "        combinations = []\n",
    "        for i in range(len(total1)):\n",
    "            for j in range(i, len(total1)):\n",
    "                if i != j:\n",
    "                    combinations.append([i, j])\n",
    "                    l = []\n",
    "                    for k in range(i, j+1):\n",
    "                        l.append(k)\n",
    "                    combinations.append(l)\n",
    "\n",
    "        b = list()\n",
    "        for sublist in combinations:\n",
    "            if sublist not in b:\n",
    "                b.append(sublist)\n",
    "        combinations = b\n",
    "\n",
    "        for i in range(len(combinations)):\n",
    "            combination = combinations[i]\n",
    "            child = parent_one.genes.tolist()\n",
    "            for j in range(len(combination)):\n",
    "                pos = combination[j]\n",
    "                interval = total1[pos]\n",
    "                reversed_list = []\n",
    "                for m in range(interval[0], interval[1]):\n",
    "                    reversed_list.append(child[m])\n",
    "                reversed_list.reverse()\n",
    "                # print(reversed_list)\n",
    "                k = 0\n",
    "                for m in range(interval[0], interval[1]):\n",
    "                    child[m] = reversed_list[k]\n",
    "                    k+=1\n",
    "\n",
    "            children.append(child)\n",
    "         \n",
    "        for i in range(len(combinations)):\n",
    "            combination = combinations[i]\n",
    "            child = parent_two.genes.tolist()\n",
    "            for j in range(len(combination)):\n",
    "                pos = combination[j]\n",
    "                interval = total2[pos]\n",
    "                reversed_list = []\n",
    "                for m in range(interval[0], interval[1]):\n",
    "                    reversed_list.append(child[m])\n",
    "                reversed_list.reverse()\n",
    "                # print(reversed_list)\n",
    "                k = 0\n",
    "                for m in range(interval[0], interval[1]):\n",
    "                    child[m] = reversed_list[k]\n",
    "                    k+=1\n",
    "\n",
    "            children.append(child)\n",
    "\n",
    "          \n",
    "        print(\"\\nChildren:\", children) \n",
    "    #************************************************************************\n",
    "        num1 = np.random.randint(0, len(children))\n",
    "        num2 = np.random.randint(0, len(children))\n",
    "        # making sure that two different children have been chosen\n",
    "        while num2 == num1: \n",
    "            num2 = np.random.randint(0, len(children))\n",
    "        child_one.genes = np.array(children[num1]) \n",
    "        child_two.genes = np.array(children[num2])\n",
    "                                                                                          \n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "chrom1 = np.array([4, 9, 7, 6, 5, 0, 8, 2, 1, 3])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = complete_subtour_exchange_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='edge-recombination'>Edge Recombination Crossover (ER)</a>\n",
    "\n",
    "In edge recombination crossover, we do the following to fill child-one.\n",
    "\n",
    "1. The first element of child-one comes from parent-one. Then, we look through the neighbors list and delete the first element of parent-one from all of the rows. Then, we look through the neighbors of the selected elements (first element of parent-one) and find the element with the least number of neighbors in the neighbors list.\n",
    "\n",
    "2. If the least number of elements for more than one element is the same, then we choose one of them randomly.\n",
    "\n",
    "3. If we reach an element in the neighbors list which has no neighbors, then we shuffle the remaining elements which are not already transferred to child-one and transfer them to it. \n",
    "\n",
    "We do the same thing for child-two with the difference that the first element of child-two will come from parent-two."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/edge_recombination_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[5 1 7 8 4 9 6 2 3 0]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[3 6 2 0 5 1 9 8 4 7]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[5 0 2 3 6 9 1 7 4 8]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[3 0 2 6 9 1 5 7 8 4]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def edge_recombination_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Edge recombination crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "  \n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        neighbour_list = []\n",
    "        for _ in range(chrom_length):\n",
    "            a = []\n",
    "            neighbour_list.append(a)\n",
    "    #******************************************************** Filling neighbour_list \n",
    "        parent_one_copy = parent_one.genes.tolist().copy()\n",
    "        parent_one_copy = sorted(parent_one_copy) #sorting the array for neighbour_list\n",
    "        # print(parent_one_copy) #sorted parent_one\n",
    "           \n",
    "        for i in range(chrom_length):\n",
    "            pos = np.where(parent_one.genes == parent_one_copy[i])\n",
    "            neighbour_list[i].append(parent_one.genes[pos[0][0] - 1])           \n",
    "            if pos[0][0] < chrom_length - 1:\n",
    "                neighbour_list[i].append(parent_one.genes[pos[0][0] + 1])\n",
    "            else:\n",
    "                neighbour_list[i].append(parent_one.genes[0])\n",
    "\n",
    "            pos = np.where(parent_two.genes == parent_one_copy[i])\n",
    "            neighbour_list[i].append(parent_two.genes[pos[0][0] - 1])           \n",
    "            if pos[0][0] < chrom_length - 1:\n",
    "                neighbour_list[i].append(parent_two.genes[pos[0][0] + 1])\n",
    "            else:\n",
    "                neighbour_list[i].append(parent_two.genes[0])\n",
    "\n",
    "        for i in range(chrom_length):\n",
    "            neighbour_list[i] = list(dict.fromkeys(neighbour_list[i]))\n",
    "        # print(neighbour_list)  \n",
    "    #***********************************************************************************       \n",
    "        neighbour_list_copy = neighbour_list.copy()\n",
    "        child1 = []\n",
    "        child2 = []\n",
    "        #%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% Filling child1\n",
    "        child1.append(parent_one.genes[0])\n",
    "        for i in range(chrom_length - 1):\n",
    "            node = child1[-1]\n",
    "            for j in range(chrom_length):\n",
    "                if node in neighbour_list_copy[j]:\n",
    "                    neighbour_list_copy[j].remove(node)\n",
    "\n",
    "            node_index = parent_one_copy.index(node)\n",
    "            # print(node_index)\n",
    "            neighbour_lengths = []\n",
    "            for j in range(len(neighbour_list_copy[node_index])): #traversing through the node looking for its remaining neighbours searching for their lengths\n",
    "                neighbour = neighbour_list_copy[node_index][j] \n",
    "                neighbour_index = parent_one_copy.index(neighbour) #finding the index of that neighbour in neighbour_list\n",
    "                neighbour_length = len(neighbour_list_copy[neighbour_index]) #calculating the length of that neighbour\n",
    "                neighbour_lengths.append(neighbour_length) #saving the length of that neighbour in the list neighbour_lengths\n",
    "                \n",
    "            # print(neighbour_list)\n",
    "            #replacing the previous node with the new node which is the neighbour of the previous node with the least length.\n",
    "            if not not neighbour_lengths: #if neighbour_lengths is not empty\n",
    "                item = np.min(neighbour_lengths)\n",
    "            else:\n",
    "                # print(\"problem\")\n",
    "                not_seen_list = []\n",
    "                for k in range(chrom_length):\n",
    "                    if parent_one.genes[k] not in child1:\n",
    "                        not_seen_list.append(parent_one.genes[k])\n",
    "                \n",
    "                np.random.shuffle(not_seen_list)\n",
    "                for k in range(len(not_seen_list)):\n",
    "                    child1.append(not_seen_list[k])\n",
    "\n",
    "                child_one.genes = np.array(child1)\n",
    "                break\n",
    "                \n",
    "            count = 0\n",
    "            for j in range(len(neighbour_lengths)): #checking whether the minimum is duplicated\n",
    "                if neighbour_lengths[j] == item:\n",
    "                    count+=1\n",
    "\n",
    "            if count == 1: #if minimum is unique, then append the node with minimum length\n",
    "                node = neighbour_list_copy[node_index][neighbour_lengths.index(np.min(neighbour_lengths))] \n",
    "            else: #if there are more than one neighbour with the least size, pick one of them randomly.\n",
    "                duplicate_elements_indexes = []\n",
    "                for j in range(len(neighbour_lengths)): #adding the indices of the duplicated element to the list duplicate_elements_indexes\n",
    "                    if neighbour_lengths[j] == item:\n",
    "                        duplicate_elements_indexes.append(j)\n",
    "                node = neighbour_list_copy[node_index][np.random.choice(duplicate_elements_indexes)]\n",
    "\n",
    "\n",
    "            child1.append(node) #appending the new node to child1\n",
    "            \n",
    "\n",
    "    #******************************************************** Filling neighbour_list          \n",
    "        for i in range(chrom_length):\n",
    "            pos = np.where(parent_one.genes == parent_one_copy[i])\n",
    "            neighbour_list[i].append(parent_one.genes[pos[0][0] - 1])           \n",
    "            if pos[0][0] < chrom_length - 1:\n",
    "                neighbour_list[i].append(parent_one.genes[pos[0][0] + 1])\n",
    "            else:\n",
    "                neighbour_list[i].append(parent_one.genes[0])\n",
    "\n",
    "            pos = np.where(parent_two.genes == parent_one_copy[i])\n",
    "            neighbour_list[i].append(parent_two.genes[pos[0][0] - 1])           \n",
    "            if pos[0][0] < chrom_length - 1:\n",
    "                neighbour_list[i].append(parent_two.genes[pos[0][0] + 1])\n",
    "            else:\n",
    "                neighbour_list[i].append(parent_two.genes[0])\n",
    "\n",
    "        for i in range(chrom_length):\n",
    "            neighbour_list[i] = list(dict.fromkeys(neighbour_list[i]))\n",
    "        # print(neighbour_list)  \n",
    "    #*********************************************************************************** \n",
    "        #%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% Filling child2\n",
    "        neighbour_list_copy = neighbour_list.copy()\n",
    "        # print(neighbour_list_copy)\n",
    "        child2.append(parent_two.genes[0])\n",
    "        for i in range(chrom_length - 1):\n",
    "            node = child2[-1]\n",
    "            for j in range(chrom_length):\n",
    "                if node in neighbour_list_copy[j]:\n",
    "                    neighbour_list_copy[j].remove(node)\n",
    "\n",
    "            node_index = parent_one_copy.index(node)\n",
    "            # print(node_index)\n",
    "            neighbour_lengths = []\n",
    "            for j in range(len(neighbour_list_copy[node_index])): #traversing through the node looking for its remaining neighbours searching for their lengths\n",
    "                neighbour = neighbour_list_copy[node_index][j] \n",
    "                neighbour_index = parent_one_copy.index(neighbour) #finding the index of that neighbour in neighbour_list\n",
    "                neighbour_length = len(neighbour_list_copy[neighbour_index]) #calculating the length of that neighbour\n",
    "                neighbour_lengths.append(neighbour_length) #saving the length of that neighbour in the list neighbour_lengths\n",
    "                \n",
    "            # print(neighbour_lengths)\n",
    "            #replacing the previous node with the new node which is the neighbour of the previous node with the least length.\n",
    "            if not not neighbour_lengths: #if neighbour_lengths is not empty\n",
    "                item = np.min(neighbour_lengths)\n",
    "            else:\n",
    "                # print(\"problem\")\n",
    "                not_seen_list = []\n",
    "                for k in range(chrom_length):\n",
    "                    if parent_one.genes[k] not in child2:\n",
    "                        not_seen_list.append(parent_one.genes[k])\n",
    "                \n",
    "                np.random.shuffle(not_seen_list)\n",
    "                for k in range(len(not_seen_list)):                   \n",
    "                    child2.append(not_seen_list[k])\n",
    "                \n",
    "                child_two.genes = np.array(child2)\n",
    "                break\n",
    "                           \n",
    "            count = 0\n",
    "            for j in range(len(neighbour_lengths)): #checking whether the minimum is duplicated\n",
    "                if neighbour_lengths[j] == item:\n",
    "                    count+=1\n",
    "\n",
    "            if count == 1: #if minimum is unique, then append the node with minimum length\n",
    "                node = neighbour_list_copy[node_index][neighbour_lengths.index(np.min(neighbour_lengths))] \n",
    "            else: #if there are more than one neighbour with the least size, pick one of them randomly.\n",
    "                duplicate_elements_indexes = []\n",
    "                for j in range(len(neighbour_lengths)): #adding the indices of the duplicated element to the list duplicate_elements_indexes\n",
    "                    if neighbour_lengths[j] == item:\n",
    "                        duplicate_elements_indexes.append(j)\n",
    "                node = neighbour_list_copy[node_index][np.random.choice(duplicate_elements_indexes)]\n",
    "\n",
    "\n",
    "            # print(count)\n",
    "            child2.append(node) #appending the new node to child1\n",
    "            child_one.genes = np.array(child1)\n",
    "            child_two.genes = np.array(child2)\n",
    "            \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([5, 1, 7, 8, 4, 9, 6, 2, 3, 0])\n",
    "chrom1 = np.array([3, 6, 2, 0, 5, 1, 9, 8, 4, 7])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = edge_recombination_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='alternate-edges'>Alternate Edges Crossover (AEX)</a>\n",
    "\n",
    "The Alternate Edges Crossover (AEX) operator works by alternating between the edges of the two parents to create a new child.To operate alternate edges crossover on two selected parents, we first need to introduce a concept named `arc`. An `arc` is an edge in a parent solution (any pair of elements that are in a sequence). For instance, the arcs for parent-one is as follows: <br>\n",
    "`5-1`, `1-7`, `7-8`, `8-4`, `4-9`, `9-6`, `6-2`, `2-3`, `3-0`, `0-5` \n",
    "\n",
    "Similarly, the arcs or edges in parent-two are: <br>\n",
    "`3-6`, `6-2`, `2-0`, `0-5`, `5-1`, `1-9`, `9-8`, `8-4`, `4-7`, `7-3`\n",
    "\n",
    "The AEX operator works by alternating between the edges of the two parents to create a new child. Now, we discuss how alternate edges crossover works with an example. Consider the following example where `parent-one = [5 1 7 8 4 9 6 2 3 0]` and `parent-two = [3 6 2 0 5 1 9 8 4 7]`.\n",
    "\n",
    "For child-one, we start by the first arc in parent-one which is `5-1`. We find the position of the element `1` in parent-two and notice that the next arc in the sequence is `1-9`. Again, we look for element `9` in parent-one and we find out that the next arc in the sequence is `9-6`. We keep alternating between the parents, until we reach a cycle (where the next element in the sequence is already seen). If we find the element `3` in parent-two, the next element in the sequence would be `6` which is already seen. So, up until now, we have the following for child-one: <br>\n",
    "`child-one = [5 1 9 6 2 3 -1 -1 -1 -1]` <br>\n",
    "where -1s denote the positions that are yet to be filled. We reached a cycle in parent-two, so we will randomly choose a value that is not already seen from the other parent which in this case is parent-one. If we choose `7` in parent-one, the corresponding arc would be `7-8` in parent-one, which then gives us the arc `8-4` in parent-two, and then the arc `4-9` in parent-one. Element `9` is already seen, therefore, we have another cycle. Thus far, we have the following for child-one: <br>\n",
    "`child-one = [5 1 9 6 2 3 7 8 4 -1]` <br>\n",
    "The only unseen element in parent-two is `0`. So, the final result for child-one would be: <br>\n",
    "`child-one = [5 1 9 6 2 3 7 8 4 0]` <br>\n",
    "\n",
    "We have a similar approach to calculate the genes of child-two, but this time, we start off by the arc `3-6` in parent-two. So, for child-two, we will have `child-two = [3 6 2 0 5 1 7 -1 -1 -1]` until we reach a cycle in parent-two (as the element `3` is already seen). If we choose the value `4` in parent-one, we get the arc `4-9` in parent-one, `9-8` in parent-two, and we are done! So, the final result for child-two will be: <br>\n",
    "`child-two = [3 6 2 0 5 1 7 4 9 8]`\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/aex.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[5 1 7 8 4 9 6 2 3 0]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[3 6 2 0 5 1 9 8 4 7]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[5 1 9 6 2 3 0 7 8 4]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[3 6 2 0 8 5 1 9 4 7]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "import collections\n",
    "\n",
    "def duplicate_list(myList):\n",
    "    \"\"\"\n",
    "    Returns True if there are no duplicate elements in the list\n",
    "    \"\"\" \n",
    "    if len(myList) == len(set(myList)):\n",
    "        return False\n",
    "    \n",
    "    return True\n",
    "\n",
    "def alternate_edges_crossover(parent_one, parent_two, pc): \n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Alternate edges crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    " \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)  \n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "               \n",
    "        #********************************************************************** For the first child\n",
    "        target_arcs = []\n",
    "        seen_list = []\n",
    "        pos = 0 # position of the next element\n",
    "        count = 0\n",
    "        alternate = 0\n",
    "        while True:\n",
    "            if count == chrom_length:\n",
    "                break\n",
    "            if alternate%2 == 0:\n",
    "                target_arcs.append(parent_one.genes[pos])\n",
    "                seen_list.append(parent_one.genes[pos])\n",
    "                if pos+1 != chrom_length:\n",
    "                    pos+=1  \n",
    "                count+=1\n",
    "                pos = parent_two.genes.tolist().index(parent_one.genes[pos]) \n",
    "                alternate+=1\n",
    "            else:\n",
    "                target_arcs.append(parent_two.genes[pos])\n",
    "                seen_list.append(parent_two.genes[pos])\n",
    "                if pos+1 != chrom_length:\n",
    "                    pos+=1  \n",
    "                count+=1\n",
    "                pos = parent_one.genes.tolist().index(parent_two.genes[pos])\n",
    "                alternate+=1\n",
    "\n",
    "\n",
    "            if count > 1:\n",
    "                n = 0\n",
    "                for k in range(len(seen_list) - 1):\n",
    "                    n+=1\n",
    "                    if target_arcs[-1] == seen_list[k]: \n",
    "                        break\n",
    "\n",
    "            not_seen_list = [] # elements that are not already transfered to the child\n",
    "            if count > 1 and n != len(seen_list) - 1: # if a cycle is found\n",
    "                target_arcs.pop()\n",
    "                for j in range(chrom_length):\n",
    "                    if parent_one.genes[j] not in seen_list:\n",
    "                        not_seen_list.append(parent_one.genes[j])\n",
    "                if not_seen_list == []:\n",
    "                    break\n",
    "\n",
    "                if alternate%2 != 0: # it means that we reached a cycle in parent_two\n",
    "                    element = not_seen_list[np.random.randint(0, len(not_seen_list))]\n",
    "                    target_arcs.append(element)\n",
    "                    seen_list.append(element)\n",
    "                    pos = parent_one.genes.tolist().index(element)\n",
    "                    if pos+1 != chrom_length:\n",
    "                        pos+=1                  \n",
    "                    pos = parent_two.genes.tolist().index(parent_one.genes[pos]) \n",
    "\n",
    "                    \n",
    "                else: # we reached a cycle in parent_one\n",
    "                    element = not_seen_list[np.random.randint(0, len(not_seen_list))]\n",
    "                    target_arcs.append(element) \n",
    "                    seen_list.append(element)\n",
    "                    pos = parent_two.genes.tolist().index(element) \n",
    "                    if pos+1 != chrom_length:\n",
    "                        pos+=1  \n",
    "                    pos = parent_one.genes.tolist().index(parent_two.genes[pos]) \n",
    "                \n",
    "        # For assurance that child_one is an authentic answer\n",
    "        #$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$ \n",
    "        duplicated_elements = [item for item, count in collections.Counter(target_arcs).items() if count > 1]\n",
    "        if duplicate_list(target_arcs):\n",
    "            not_seen_list = []\n",
    "            for i in range(chrom_length): #filling the not_seen_list \n",
    "                if parent_one.genes[i] not in target_arcs:\n",
    "                    not_seen_list.append(parent_one.genes[i])\n",
    "            # replacing the duplicated elements with the elements that are not seen\n",
    "            for i in range(len(duplicated_elements)): \n",
    "                location = len(target_arcs) - 1 - target_arcs[::-1].index(duplicated_elements[i]) \n",
    "                target_arcs[location] = not_seen_list.pop() # last duplicated element\n",
    "        #$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$\n",
    "        child_one.genes = np.array(target_arcs)\n",
    "\n",
    "        # target_arcs = list(dict.fromkeys(target_arcs))\n",
    "        # seen_list = list(dict.fromkeys(seen_list))\n",
    "        #************************************************************************************ For the second child\n",
    "        target_arcs = []\n",
    "        seen_list = []\n",
    "        pos = 0 # position of the next element\n",
    "        count = 0\n",
    "        alternate = 1\n",
    "        while True:\n",
    "            if count == chrom_length:\n",
    "                break\n",
    "            if alternate%2 == 0:\n",
    "                target_arcs.append(parent_one.genes[pos])\n",
    "                seen_list.append(parent_one.genes[pos])\n",
    "                if pos+1 != chrom_length:\n",
    "                    pos+=1  \n",
    "                count+=1\n",
    "                pos = parent_two.genes.tolist().index(parent_one.genes[pos]) \n",
    "                alternate+=1\n",
    "            else:\n",
    "                target_arcs.append(parent_two.genes[pos])\n",
    "                seen_list.append(parent_two.genes[pos])\n",
    "                if pos+1 != chrom_length:\n",
    "                    pos+=1  \n",
    "                count+=1\n",
    "                pos = parent_one.genes.tolist().index(parent_two.genes[pos])\n",
    "                alternate+=1\n",
    "\n",
    "            if count > 1:\n",
    "                n = 0\n",
    "                for k in range(len(seen_list) - 1):\n",
    "                    n+=1\n",
    "                    if target_arcs[-1] == seen_list[k]: \n",
    "                        break\n",
    "\n",
    "            not_seen_list = [] # elements that are not already transfered to child\n",
    "            if count > 1 and n != len(seen_list) - 1: # if a cycle is found\n",
    "                target_arcs.pop()\n",
    "                for j in range(chrom_length):\n",
    "                    if parent_one.genes[j] not in seen_list:\n",
    "                        not_seen_list.append(parent_one.genes[j])\n",
    "                if not_seen_list == []:\n",
    "                    break\n",
    "\n",
    "                if alternate%2 != 0: #it means that we reached a cycle in parent_two\n",
    "                    element = not_seen_list[np.random.randint(0, len(not_seen_list))]\n",
    "                    target_arcs.append(element)\n",
    "                    seen_list.append(element)\n",
    "                    pos = parent_one.genes.tolist().index(element)\n",
    "                    if pos+1 != chrom_length:\n",
    "                        pos+=1                  \n",
    "                    pos = parent_two.genes.tolist().index(parent_one.genes[pos]) \n",
    "\n",
    "                    \n",
    "                else: # we reached a cycle in parent_one\n",
    "                    element = not_seen_list[np.random.randint(0, len(not_seen_list))]\n",
    "                    target_arcs.append(element) \n",
    "                    seen_list.append(element)\n",
    "                    pos = parent_two.genes.tolist().index(element) \n",
    "                    if pos+1 != chrom_length:\n",
    "                        pos+=1  \n",
    "                    pos = parent_one.genes.tolist().index(parent_two.genes[pos]) \n",
    "                \n",
    "        # For assurance that child_two is an authentic answer\n",
    "        #$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$ \n",
    "        duplicated_elements = [item for item, count in collections.Counter(target_arcs).items() if count > 1]\n",
    "        if duplicate_list(target_arcs):\n",
    "            not_seen_list = []\n",
    "            for i in range(chrom_length): #filling the not_seen_list \n",
    "                if parent_one.genes[i] not in target_arcs:\n",
    "                    not_seen_list.append(parent_one.genes[i])\n",
    "            # replacing the duplicated elements with the elements that are not seen\n",
    "            for i in range(len(duplicated_elements)): \n",
    "                location = len(target_arcs) - 1 - target_arcs[::-1].index(duplicated_elements[i]) \n",
    "                target_arcs[location] = not_seen_list.pop() # last duplicated element\n",
    "        #$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$\n",
    "        child_two.genes = np.array(target_arcs)\n",
    "      \n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([5, 1, 7, 8, 4, 9, 6, 2, 3, 0])\n",
    "chrom1 = np.array([3, 6, 2, 0, 5, 1, 9, 8, 4, 7])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = alternate_edges_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='half-uniform'>Half Uniform Crossover (HUX)</a>\n",
    "\n",
    "Half uniform crossover (HUX), is an extension of the uniform crossover (UX). In this crossover, we first copy the genes of parent-one into child-one and the genes of parent-two into child-two. Then, we loop over child-one and child-two and wherever the value of the gene `i` for child-one is not equal to the value of the gene `i` for child-two, we produce a uniform random real number `w` in the interval `[0, 1]` and do the following for that gene: <br> \n",
    "If `w` is less than or equal to 0.5, then we swap the value of the gene in the two children. Otherwise, no change will be made to the value of that gene in the children."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/hux.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0 1 0 0 1 1 0 1 1 0 1]\n",
      "ID=#1, Fitness=7, \n",
      "Genes=\n",
      "[1 0 1 1 0 1 0 1 1 0 1]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0 0 0 1 1 1 0 1 1 0 1]\n",
      "ID=#1, Fitness=7, \n",
      "Genes=\n",
      "[1 1 1 0 0 1 0 1 1 0 1]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def half_uniform_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        # copying parents to their children\n",
    "        child_one.genes = parent_one.genes\n",
    "        child_two.genes = parent_two.genes\n",
    "\n",
    "        for i in range(chrom_length):\n",
    "            if child_one.genes[i] != child_two.genes[i]:\n",
    "                w = np.random.uniform(low=0, high=1)\n",
    "                if w <= 0.5: # swapping bits\n",
    "                    child_one.genes[i], child_two.genes[i] = child_two.genes[i], child_one.genes[i]  \n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "    \n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1])\n",
    "chrom1 = np.array([1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = half_uniform_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  # two children\n",
    "        Chromosome.describe(CROSS[index])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='highly-disruptive'>Highly Disruptive Crossover (HDX)</a>\n",
    "\n",
    "Highly Disruptive Crossover is a type of crossover operator that is designed to create a high degree of diversity in the offspring population. It is achieved by randomly selecting a subset of genes from each parent and swapping them between the parents. \n",
    "\n",
    "Here's an example of Highly Disruptive Crossover:\n",
    "\n",
    "Let's say we have two parent chromosomes: `10101010` and `11001100`. We randomly select a subset of genes from each parent, say the first two genes from the first parent and the last two genes from the second parent. The resulting offspring will be `00101010` and `11001110`. As you can see, the offspring has a combination of genes from both parents, but the order of the genes has been disrupted. This helps to create a high degree of diversity in the offspring population.\n",
    "\n",
    "In Highly Disruptive Crossover, the subset of genes from each parent that are swapped do not need to be adjacent to each other. They can be separated by any number of genes. The only requirement is that the subset of genes selected from each parent should be of the same length.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/hdx.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=0, \n",
      "Genes=\n",
      "[0 0 0 0 0 0 0 0 0 0 0]\n",
      "ID=#1, Fitness=11, \n",
      "Genes=\n",
      "[1 1 1 1 1 1 1 1 1 1 1]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[1 0 0 1 1 0 1 0 1 1 0]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[0 1 0 1 1 1 1 0 0 0 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def highly_disruptive_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([-1]*chrom_length),id_=1,fitness = 125.2)\n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        # copying parents to their children\n",
    "        child_one.genes = parent_one.genes\n",
    "        child_two.genes = parent_two.genes\n",
    "\n",
    "        # Defining the arrays (children genes)\n",
    "        arr1 = child_one.genes\n",
    "        arr2 = child_two.genes\n",
    "\n",
    "        # Number of genes to swap\n",
    "        swap_num = np.random.randint(1, chrom_length)\n",
    "\n",
    "        # Generate random indices to swap for each array\n",
    "        indices1 = np.random.choice(len(arr1), size=swap_num, replace=False)\n",
    "        indices2 = np.random.choice(len(arr2), size=swap_num, replace=False)\n",
    "\n",
    "        # Swap the values at the selected indices for each array\n",
    "        arr1[indices1], arr2[indices2] = arr2[indices2], arr1[indices1]\n",
    " \n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "    \n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])\n",
    "chrom1 = np.array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = highly_disruptive_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  # two children\n",
    "        Chromosome.describe(CROSS[index])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='blend'>Blend Crossover (BLX)</a>\n",
    "\n",
    "To perform blend crossover, we need three parameters. \n",
    "- A list named `d` which will calculate the distance between the gene values of the parents. \n",
    "- A positive real parameter named `a` in the interval [0, 1] which is selected randomly before the operation. \n",
    "- Then for each gene of the children, we create a random real uniform number `u` for each offspring. The value of the gene in that position will become this random real uniform number. `u` is calculated as the following:\n",
    "\n",
    "> u ~ ( min(parent_one[i], parent_two[i]) - (a * d[i]), max(parent_one[i], parent_two[i]) + (a * d[i]) )."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/blend_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.121 0.152 0.231 0.143 0.732 0.315 0.434 0.633]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[0.765 0.168 0.914 0.435 0.893 0.332 0.981 0.194]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0.125 0.151 0.701 0.093 0.852 0.32  0.909 0.202]\n",
      "ID=#1, Fitness=7, \n",
      "Genes=\n",
      "[0.334 0.165 0.347 0.255 0.755 0.33  0.303 0.18 ]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy \n",
    "\n",
    "def blend_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Blend crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0.0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([0.0]*chrom_length),id_=1,fitness = 125.2)\n",
    "  \n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        distance_vector = []\n",
    "        for i in range(chrom_length):\n",
    "            distance_vector.append(abs(parent_one.genes[i] - parent_two.genes[i]))\n",
    "        # print(distance_vector)\n",
    "        a = np.random.uniform(low = 0, high = 1)\n",
    "        for i in range(chrom_length):\n",
    "            w = np.random.uniform(low = abs(np.minimum(parent_one.genes[i], parent_two.genes[i]) - a * distance_vector[i]), high = abs(np.maximum(parent_one.genes[i], parent_two.genes[i]) + a * distance_vector[i]))\n",
    "            child_one.genes[i] = round(w, 3) \n",
    "            w = np.random.uniform(low = abs(np.minimum(parent_one.genes[i], parent_two.genes[i]) - a * distance_vector[i]), high = abs(np.maximum(parent_one.genes[i], parent_two.genes[i]) + a * distance_vector[i]))\n",
    "            child_two.genes[i] = round(w, 3)\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0.121, 0.152, 0.231, 0.143, 0.732, 0.315, 0.434, 0.633])\n",
    "chrom1 = np.array([0.765, 0.168, 0.914, 0.435, 0.893, 0.332, 0.981, 0.194])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = blend_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='linear-recombination'>Linear Recombination Crossover</a>\n",
    "\n",
    "In linear recombination crossover, we select two parents and create three offspring. If we name the first parent as parent-one and the second parent as parent-two, the three offspring are calculated as follow: <br>\n",
    "<span style=\"color:#3578EE\">**First offspring:**</span> `parent_one + parent_two` <br>\n",
    "<span style=\"color:#EE6897\">**Second offspring:**</span> `(1.5)*parent_one - (0.5)*parent_two` <br>\n",
    "<span style=\"color:#25EE97\">**Third offspring:**</span> `(-0.5)*parent_one + (1.5)*parent_two` <br>\n",
    "\n",
    "After the production of the offspring, we choose the two fittest offspring and return them as the result."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/linear_recombination_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.121 0.152 0.231 0.143 0.732 0.315 0.434 0.633]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[0.765 0.168 0.914 0.435 0.893 0.332 0.981 0.194]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#1, Fitness=6, \n",
      "Genes=\n",
      "[0.201  0.144  0.1105 0.003  0.6515 0.3065 0.1605 0.8525]\n",
      "ID=#2, Fitness=3, \n",
      "Genes=\n",
      "[1.087  0.176  1.2555 0.581  0.9735 0.3405 1.2545 0.0255]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy \n",
    "\n",
    "def linear_recombination_crossover(parent_one, parent_two, pc): \n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Linear crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one) \n",
    "    child_one = Chromosome(genes= np.array([0.0]*chrom_length),id_=0,fitness = 130.3)   \n",
    "    child_two = Chromosome(genes= np.array([0.0]*chrom_length),id_=1,fitness = 129.2)\n",
    "    child_three = Chromosome(genes= np.array([0.0]*chrom_length),id_=2,fitness = 129.3)\n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "\n",
    "        child_one.genes = abs(parent_one.genes + parent_two.genes)\n",
    "        child_two.genes = abs((1.5)*parent_one.genes - (0.5)*parent_two.genes)\n",
    "        child_three.genes = abs((-0.5)*parent_one.genes + (1.5)*parent_two.genes)\n",
    "\n",
    "\n",
    "        children = [] # List containing the children\n",
    "        children.append(child_one)\n",
    "        children.append(child_two)\n",
    "        children.append(child_three)\n",
    "\n",
    "        fitnesses = [] # List containing the fitnesses\n",
    "        fitnesses.append(fitness_function(child_one.genes))\n",
    "        fitnesses.append(fitness_function(child_two.genes))\n",
    "        fitnesses.append(fitness_function(child_three.genes))\n",
    "\n",
    "        # removing the child with the worth fitness\n",
    "        minimum_fitness_index = fitnesses.index(np.min(fitnesses))\n",
    "        children.pop(minimum_fitness_index) \n",
    "\n",
    "\n",
    "        child_one = children[0]\n",
    "        child_two = children[1]\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "    \n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([0.121, 0.152, 0.231, 0.143, 0.732, 0.315, 0.434, 0.633])\n",
    "chrom1 = np.array([0.765, 0.168, 0.914, 0.435, 0.893, 0.332, 0.981, 0.194])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = linear_recombination_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='directional-heuristic'>Directional Heuristic Crossover</a>\n",
    "\n",
    "In directional heuristic crossover, for gene `i` of the child, we have: <br>\n",
    "`child-one[i] = U(0, 1)*(parent-two[i] - parent-one[i]) + parent-two[i]` <br>\n",
    "where U(0, 1) is a random real number between 0 and 1.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/dhx.png\"  alt='missing' width=\"600\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[0.765 0.168 0.914 0.435 0.893 0.332 0.981 0.194]\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.121 0.152 0.231 0.143 0.732 0.315 0.434 0.633]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.193 0.141 0.144 0.003 0.686 0.312 0.317 0.926]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def directional_heuristic_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Directional heuristic crossover on them.\n",
    "    This crossover is subject to the constraint that the parent_two cannot be worth than parent_one. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    # if parent_one is better, switch the parents\n",
    "    if parent_one.fitness > parent_two.fitness: \n",
    "        parent_one, parent_two = parent_two, parent_one\n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0.0]*chrom_length), id_=0, fitness = 125.2)  # child\n",
    "  \n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        for i in range(chrom_length):\n",
    "            child_one.genes[i] = abs(round(np.random.uniform(low=0.0, high=1.0)*(parent_two.genes[i] - parent_one.genes[i]) + parent_two.genes[i], 3))\n",
    "            \n",
    "        \n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        \n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "\n",
    "    return child_one\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0.121, 0.152, 0.231, 0.143, 0.732, 0.315, 0.434, 0.633])\n",
    "chrom1 = np.array([0.765, 0.168, 0.914, 0.435, 0.893, 0.332, 0.981, 0.194])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = directional_heuristic_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "\n",
    "    Chromosome.describe(CROSS)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='geometrical'>Geometrical Crossover</a>\n",
    "\n",
    "In geometrical crossover, one offspring is produced, and the value of its genes is calculated by the geometrical average of the same genes in parents. If we have `n` parents, each gene of the offspring will be produced as follows: \n",
    "\n",
    "`child-one = (parent_one)^(a_1) * (parent_two)^(a_2) *  * (parent_n)^(a_n)`\n",
    "\n",
    "where $_{i=1}^{n} a_i = 1$. This crossover can also be used in multi-parent recombination."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/geometrical_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.121 0.152 0.231 0.143 0.732 0.315 0.434 0.633]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[0.765 0.168 0.914 0.435 0.893 0.332 0.981 0.194]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0.371 0.162 0.533 0.281 0.826 0.325 0.712 0.309]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "import math\n",
    "\n",
    "def geometrical_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs geometrical crossover on them.\n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0.0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "  \n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "\n",
    "        alpha1 = np.random.uniform(low=0.0, high=0.5)  # alpha1 + alpha2 = 1\n",
    "        alpha2 = 1 - alpha1   \n",
    "\n",
    "        for i in range(chrom_length):\n",
    "            child_one.genes[i] = round((parent_one.genes[i] ** alpha1) * (parent_two.genes[i] ** alpha2), 3)\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0.121, 0.152, 0.231, 0.143, 0.732, 0.315, 0.434, 0.633])\n",
    "chrom1 = np.array([0.765, 0.168, 0.914, 0.435, 0.893, 0.332, 0.981, 0.194])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = geometrical_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(CROSS)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='simulated-binary'>Simulated Binary Crossover (SBX)</a>\n",
    "\n",
    "Simulated Binary Crossover (SBX) is a crossover operator used in genetic algorithms. It is designed to work with real-valued chromosomes. The SBX operator works by simulating the binary crossover operator used in binary-coded GAs. The idea is to retain the property of binary coding with single-point crossover, which states that the average of the parent chromosomes is equal to the average of the child chromosomes. In SBX, this property is retained by using a probability distribution that simulates the binary crossover. The crossover operator selects two parent chromosomes and generates two offspring chromosomes by applying the SBX operator. The SBX operator works by generating a random number between 0 and 1, and then using this number to calculate the beta value. The beta value is then used to calculate the offspring chromosomes from the parent chromosomes.\n",
    "\n",
    "Here is an example of how SBX works:\n",
    "\n",
    "Let's say we have two parent chromosomes: `p1 = [1.5, 2.0, 3.0]` and `p2 = [2.5, 3.0, 4.0]`. We randomly select a beta value between 0 and 1, say `beta = 0.5`. The resulting offspring chromosomes will be:\n",
    "\n",
    "```\n",
    "c1 = [1.75, 2.5, 3.5]\n",
    "c2 = [2.25, 2.5, 3.5]\n",
    "```\n",
    "\n",
    "As you can see, the offspring chromosomes have a combination of genes from both parents, but the values have been modified based on the beta value.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/simulated_binary_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.121 0.152 0.231 0.143 0.732 0.315 0.434 0.633]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[0.765 0.168 0.914 0.435 0.893 0.332 0.981 0.194]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.14  0.15  0.201 0.143 0.731 0.316 0.371 0.637]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[0.746 0.17  0.944 0.435 0.894 0.331 1.044 0.19 ]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def simulated_binary_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Simulated binary crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0.0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([0.0]*chrom_length),id_=1,fitness = 125.2)\n",
    "  \n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number      \n",
    "        distribution_index = np.random.randint(1, chrom_length) #distribution_index > 0\n",
    "        factor = 0\n",
    "        for i in range(chrom_length):\n",
    "            random_number = np.random.uniform(low=0.0, high=1.0)\n",
    "            if random_number <= 0.5:\n",
    "                factor = (2 * random_number) ** (1 / (distribution_index + 1))\n",
    "            else:\n",
    "                factor = (1 / (2 * (1 - random_number))) ** (1 / (distribution_index + 1))\n",
    "\n",
    "            child_one.genes[i] = round(abs(0.5 * ((1 + factor) * parent_one.genes[i] + (1 - factor) * parent_two.genes[i])), 3)   \n",
    "            child_two.genes[i] = round(abs(0.5 * ((1 - factor) * parent_one.genes[i] + (1 + factor) * parent_two.genes[i])), 3)  \n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0.121, 0.152, 0.231, 0.143, 0.732, 0.315, 0.434, 0.633])\n",
    "chrom1 = np.array([0.765, 0.168, 0.914, 0.435, 0.893, 0.332, 0.981, 0.194])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = simulated_binary_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='hill-climbing'>Hill-climbing Crossover</a>\n",
    "\n",
    "Hill-climbing is a local search algorithm that starts with an initial solution and iteratively improves the solution by making small modifications to it. Hill-climbing crossover is a variation of hill-climbing that is used in genetic algorithms. In hill-climbing crossover, the algorithm starts with an initial population of solutions, and then selects pairs of solutions to be crossed over. In hill-climbing crossover, we use uniform crossover to produce offspring; but this time, we keep doing it until an offspring is produced which has a better fitness than one of her parents.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/hill_cross.png\"  alt='missing' width=\"600\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 0 0 1 0 0 1 1 1]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[1 0 1 0 0 1 0 1 0 0]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=3, \n",
      "Genes=\n",
      "[0 1 0 0 0 0 0 1 1 0]\n",
      "ID=#1, Fitness=6, \n",
      "Genes=\n",
      "[1 0 1 0 1 1 0 1 0 1]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def claculate_mask(chrom_size, px):\n",
    "    \"\"\"\n",
    "    This method, performs the mask calculation for the Uniform crossover.\n",
    "    chrom_size: The mask has the same size of the chromosome.\n",
    "    px: A control parameter for the number of ones in the mask vector.\n",
    "    \"\"\"  \n",
    "    mask = [0]*chrom_size\n",
    "    for i in range(chrom_size):\n",
    "        prob = np.random.rand()\n",
    "        if prob <= px:\n",
    "            mask[i]=1\n",
    "    return mask\n",
    "            \n",
    "\n",
    "def uniform_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    This function takes two parents, and performs Uniform crossover on them. \n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([0]*chrom_length),id_=1,fitness = 125.2)\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        mask = claculate_mask(chrom_length,np.random.uniform(low=0.2, high=0.85))\n",
    "        for i in range(chrom_length):\n",
    "            if mask[i]==1:\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "                child_two.genes[i] = parent_two.genes[i]\n",
    "            else:\n",
    "                child_one.genes[i] = parent_two.genes[i]\n",
    "                child_two.genes[i] = parent_one.genes[i]\n",
    "           \n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "def hillclimbing_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    parent_one = Chromosome(genes= np.array([0, 1, 0, 0, 1, 0, 0, 1, 1, 1]),id_=0,fitness = 5)   \n",
    "    parent_two = Chromosome(genes= np.array([1, 0, 1, 0, 0, 1, 0, 1, 0, 0]),id_=1,fitness = 4) #repair fitness_function\n",
    "    \n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([0]*chrom_length),id_=1,fitness = 125.2)\n",
    "  \n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number\n",
    "        maximum_parent_fitness = np.maximum(parent_one.fitness, parent_two.fitness)\n",
    "        child_one, child_two = uniform_crossover(parent_one, parent_two, 1)\n",
    "        child_one.fitness = fitness_function(child_one.genes)\n",
    "        child_two.fitness = fitness_function(child_two.genes)\n",
    "        maximum_child_fitness = np.maximum(child_one.fitness, child_two.fitness)\n",
    "        \n",
    "        # Keep going until one of the children becomes better than parents\n",
    "        if maximum_child_fitness <= maximum_parent_fitness:\n",
    "            while maximum_child_fitness <= maximum_parent_fitness: \n",
    "                child_one, child_two = uniform_crossover(child_one, child_two, 1)\n",
    "                child_one.fitness = fitness_function(child_one.genes)\n",
    "                child_two.fitness = fitness_function(child_two.genes)\n",
    "                maximum_child_fitness = np.maximum(child_one.fitness, child_two.fitness)\n",
    "\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "    \n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 0, 0, 1, 0, 0, 1, 1, 1])\n",
    "chrom1 = np.array([1, 0, 1, 0, 0, 1, 0, 1, 0, 0])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = hillclimbing_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='arithmetic'>Arithmetic Crossover</a>\n",
    "\n",
    "Arithmetic Crossover is a type of crossover operator used in genetic algorithms. It is designed to work with real-valued chromosomes. The arithmetic crossover operator works by generating a new offspring chromosome by taking a weighted average of the corresponding genes in the parent chromosomes. The weight is determined by a random number between 0 and 1, and the offspring chromosome is generated as follows:\n",
    "\n",
    "```\n",
    "c[i] = alpha * p1[i] + (1 - alpha) * p2[i]\n",
    "```\n",
    "\n",
    "where `c[i]` is the `i`-th gene in the offspring chromosome, `p1[i]` and `p2[i]` are the corresponding genes in the parent chromosomes, and `alpha` is the weight factor. The value of `alpha` is usually chosen randomly from a uniform distribution between 0 and 1.\n",
    "\n",
    "Here's an example of how arithmetic crossover works:\n",
    "\n",
    "Let's say we have two parent chromosomes: `p1 = [1.5, 2.0, 3.0]` and `p2 = [2.5, 3.0, 4.0]`. We randomly select a weight factor `alpha = 0.5`. The resulting offspring chromosome will be:\n",
    "\n",
    "```\n",
    "c = [2.0, 2.5, 3.5]\n",
    "```\n",
    "\n",
    "As you can see, the offspring chromosome has a combination of genes from both parents, but the values have been modified based on the weight factor.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Crossover/arithmetic_crossover.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parents\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.121 0.152 0.231 0.143 0.732 0.315 0.434 0.633]\n",
      "ID=#1, Fitness=4, \n",
      "Genes=\n",
      "[0.765 0.168 0.914 0.435 0.893 0.332 0.981 0.194]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[0.592 0.164 0.731 0.357 0.85  0.327 0.834 0.312]\n",
      "ID=#1, Fitness=5, \n",
      "Genes=\n",
      "[0.294 0.156 0.414 0.221 0.775 0.32  0.581 0.515]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def arithmetic_crossover(parent_one, parent_two, pc):\n",
    "    \"\"\"\n",
    "    parent_one: First parent\n",
    "    parent_two: Second parent\n",
    "    pc: The probability of crossover\n",
    "    \"\"\"  \n",
    "\n",
    "    print(\"\\nParents\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    Chromosome.describe(parent_two)\n",
    "    \n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0.0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    child_two = Chromosome(genes= np.array([0.0]*chrom_length),id_=1,fitness = 125.2)\n",
    "  \n",
    "\n",
    "    if np.random.rand() < pc:  # if pc is greater than random number  \n",
    "        alpha = np.random.uniform(low=0.0, high=1.0)\n",
    "        for i in range(chrom_length):\n",
    "            child_one.genes[i] = round((1 - alpha) * parent_one.genes[i] + alpha * parent_two.genes[i], 3)\n",
    "            child_two.genes[i] = round(alpha * parent_one.genes[i] + (1 - alpha) * parent_two.genes[i], 3)    \n",
    "\n",
    "\n",
    "    else:  # if pc is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "        child_two = deepcopy(parent_two)\n",
    "    \n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    child_two.fitness = fitness_function(child_two.genes)\n",
    "\n",
    "    return child_one, child_two\n",
    "\n",
    "\n",
    "chrom0 = np.array([0.121, 0.152, 0.231, 0.143, 0.732, 0.315, 0.434, 0.633])\n",
    "chrom1 = np.array([0.765, 0.168, 0.914, 0.435, 0.893, 0.332, 0.981, 0.194])\n",
    "parent_one = Chromosome(chrom0, id_= 0, fitness = fitness_function(chrom0))  \n",
    "parent_two = Chromosome(chrom1, id_= 1, fitness = fitness_function(chrom1))\n",
    "CROSS = arithmetic_crossover(parent_one, parent_two, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    for index, _ in enumerate(CROSS):  #two children\n",
    "        Chromosome.describe(CROSS[index])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "1. Engelbrecht, A. P. (2007). Computational intelligence: An introduction (2nd ed.). John Wiley & Sons.\n",
    "\n",
    "2. Crossover (genetic algorithm). (n.d.). In Wikipedia. Retrieved December 27, 2023, from https://en.wikipedia.org/wiki/Crossover_%28genetic_algorithm%29\n",
    "\n",
    "3. Genetic Algorithm: A Learning Experience. (n.d.). In UNSW Sites. Retrieved December 27, 2023, from https://www.cse.unsw.edu.au/~cs9417ml/GA2/crossover_recom.html\n",
    "\n",
    "4. Crossover, edge recombination operator, and fly algorithm. (n.d.). In INDIAai. Retrieved December 27, 2023, from https://indiaai.gov.in/article/crossover-edge-recombination-operator-and-fly-algorithm\n",
    "\n",
    "5. Crossover in Genetic Algorithm. (n.d.). In GeeksforGeeks. Retrieved December 27, 2023, from https://www.geeksforgeeks.org/crossover-in-genetic-algorithm/\n",
    "\n",
    "6. Easiest way to implement cycle crossover. (n.d.). In Code Review Stack Exchange. Retrieved December 27, 2023, from https://codereview.stackexchange.com/questions/226179/easiest-way-to-implement-cycle-crossover\n",
    "\n",
    "7. Partially Mapped Crossover in Genetic Algorithms. (n.d.). In Baeldung. Retrieved December 27, 2023, from https://www.baeldung.com/cs/ga-pmx-operator\n",
    "\n",
    "8. pmx: Partially Mapped Crossover. (n.d.). In R Package Documentation. Retrieved December 27, 2023, from https://rdrr.io/cran/adana/man/pmx.html\n",
    "\n",
    "9. Bing, M. (2023). Recombination in genetic algorithms [Graphic art]. Image Creator from Microsoft Designer.\n",
    "\n",
    "10. cpc: Count-preserving Crossover (CPC). (n.d.). In R Package Documentation. Retrieved December 27, 2023, from https://rdrr.io/cran/adana/man/cpc.html.\n",
    "\n",
    "11. Gwiazda, T. (n.d.). Count-preserving Crossover. Retrieved December 27, 2023, from http://www.tomaszgwiazda.com/countPX.htm\n",
    "\n",
    "12. Leps, M. (n.d.). Genetic Algorithms. Retrieved December 27, 2023, from https://mech.fsv.cvut.cz/~leps/teaching/mom/lectures/lecture06_GA.pdf\n",
    "\n",
    "13. Crossover (genetic algorithm). (n.d.). In bionity.com. Retrieved December 27, 2023, from https://www.bionity.com/en/encyclopedia/Crossover_%28genetic_algorithm%29.html\n",
    "\n",
    "14. Gwiazda, T. (n.d.). Introduction to: Genetic algorithms reference Volume I Crossover for .... Retrieved December 27, 2023, from http://tomaszgwiazda.com/wstep.htm\n",
    "\n",
    "15. Kaur, H., & Singh, S. (2017). Multi-objective optimisation of reliable product-plant network .... Applied Network Science, 2(1), 1-16. https://doi.org/10.1007/s41109-017-0058-8\n",
    "\n",
    "16. Srinivas, M., & Patnaik, L. M. (2022). A Study of Crossover Operators in Genetic Algorithms. In L. M. Patnaik, & M. Srinivas (Eds.), Handbook of Genetic Algorithms (pp. 23-45). Springer. https://doi.org/10.1007/978-981-16-3128-3_2\n",
    "\n",
    "17. Goldberg, D. E. (2014). Genetic Algorithms. In J. Sammut, & G. I. Webb (Eds.), Encyclopedia of Machine Learning and Data Mining (pp. 487-491). Springer. https://doi.org/10.1007/978-3-319-07124-4_28\n",
    "\n",
    "18. Srinivas, M., & Patnaik, L. M. (2022). A Study of Crossover Operators in Genetic Algorithms. In L. M. Patnaik, & M. Srinivas (Eds.), Handbook of Genetic Algorithms (pp. 23-45). Springer. https://doi.org/10.1007/978-981-16-3128-3_2\n",
    "\n",
    "19. Aegis Software. (n.d.). The Impact of Disruptive Industrial Crossover. Retrieved December 27, 2023, from https://www.aiscorp.com/blog/blurred-and-disrupted-the-impact-of-industry-crossover/\n",
    "\n",
    "20. Goldberg, D. E. (2014). Genetic Algorithms. In J. Sammut, & G. I. Webb (Eds.), Encyclopedia of Machine Learning and Data Mining (pp. 487-491). Springer. https://doi.org/10.1007/978-3-319-07124-4_28\n",
    "\n",
    "21. Unequal crossing over. (n.d.). In Wikipedia. Retrieved December 27, 2023, from https://en.wikipedia.org/wiki/Unequal_crossing_over\n",
    "\n",
    "22. University of Texas at Austin. (2018, July 9). Jumping genes: Cross species transfer of genes has driven evolution. ScienceDaily. Retrieved December 27, 2023, from https://www.sciencedaily.com/releases/2018/07/180709101216.htm\n",
    "\n",
    "23. Srinivas, M., & Patnaik, L. M. (2022). A Study of Crossover Operators in Genetic Algorithms. In L. M. Patnaik, & M. Srinivas (Eds.), Handbook of Genetic Algorithms (pp. 23-45). Springer. https://doi.org/10.1007/978-981-16-3128-3_2\n",
    "\n",
    "24. Szkely, G. J., & Rizzo, M. L. (2021). Approximate distance correlation for selecting highly informative variables. PLoS Computational Biology, 17(11), e1009548. https://doi.org/10.1371/journal.pcbi.1009548\n",
    "\n",
    "25. Deb, K., & Agrawal, S. (2003). Simulated Binary Crossover for Continuous Search Space. Complex Systems, 15(3), 255-282. https://doi.org/10.1142/S0129183103005783\n",
    "\n",
    "26. Deb, K. (2001). Simulated binary crossover for continuous search space. Complex Systems, 13(6), 493-511. https://doi.org/10.1007/s00453-002-0974-z\n",
    "\n",
    "27. Stack Overflow. (2014, March 17). Simulated Binary Crossover (SBX) crossover operator example. Retrieved December 27, 2023, from https://stackoverflow.com/questions/22457941/simulated-binary-crossover-sbx-crossover-operator-example\n",
    "\n",
    "28. Sudhoff, S. D. (n.d.). Real-Coded Genetic Algorithms. Retrieved December 27, 2023, from https://engineering.purdue.edu/~sudhoff/ee630/Lecture04.pdf\n",
    "\n",
    "29. Paskorn, T. (n.d.). Simulated Binary Crossover. Retrieved December 27, 2023, from https://www.slideshare.net/paskorn/simulated-binary-crossover-presentation\n",
    "\n",
    "30. undefined. (n.d.). Retrieved December 27, 2023, from http://www.slideshare.net/\n",
    "\n",
    "31. University of Massachusetts Boston. (n.d.). DSSG. Retrieved December 27, 2023, from http://dssg.cs.umb.edu\n",
    "\n",
    "32. mcga. (n.d.). sbx_crossover: Performs sbx (simulated binary) crossover operation on a .... Retrieved December 27, 2023, from https://rdrr.io/cran/mcga/man/sbx_crossover.html\n",
    "\n",
    "33. pymoo. (n.d.). Crossover. Retrieved December 27, 2023, from https://pymoo.org/operators/crossover.html\n",
    "\n",
    "34. Kuo, T. T., & Lin, C. H. (2011). Hill-Climbing search and diversification within an evolutionary algorithm for feature selection. BioData Mining, 4(1), 23. https://doi.org/10.1186/1756-0381-4-23\n",
    "\n",
    "35. Stack Overflow. (2016, Nov 1). How is a genetic algorithm with only selection and mutation same as a hill climb? Retrieved December 27, 2023, from https://stackoverflow.com/questions/40334711/how-is-a-genetic-algorithm-with-only-selection-and-mutation-same-as-a-hill-climb\n",
    "\n",
    "36. Testbook. (n.d.). Genetic Algorithms MCQ [Free PDF]. Retrieved December 27, 2023, from https://testbook.com/objective-questions/mcq-on-genetic-algorithms--5eea6a0e39140f30f369e524\n",
    "\n",
    "37. Hansen, N. (2015). An Analysis of Integration of Hill Climbing in Crossover and Mutation Operators of Evolution Strategies. In Proceedings of the 2015 Annual Conference on Genetic and Evolutionary Computation (pp. 209-216). ACM. https://doi.org/10.1145/2739480.2754773\n",
    "\n",
    "38. Thierens, D., & Bosman, P. A. (2012). Real-Coded Memetic Algorithms with Crossover Hill-Climbing. Evolutionary Computation, 20(2), 273-300. https://doi.org/10.1162/EVCO_a_00053\n",
    "\n",
    "39. Garg, A. (2021, August 16). Crossover Operators in Genetic Algorithm. Medium. Retrieved December 27, 2023, from https://medium.com/geekculture/crossover-operators-in-ga-cffa77cdd0c8\n",
    "\n",
    "40. Crossover (genetic algorithm). (n.d.). In Wikipedia. Retrieved December 27, 2023, from https://en.wikipedia.org/wiki/Crossover_%28genetic_algorithm%29\n",
    "\n",
    "41. Scassellati, B. (n.d.). Genetic Algorithms. Retrieved December 27, 2023, from https://zoo.cs.yale.edu/classes/cs470/lectures/s2019/17-Genetic-Algorithms.pdf\n",
    "\n",
    "42. Tutorialspoint. (n.d.). Genetic Algorithms - Crossover. Retrieved December 27, 2023, from https://www.tutorialspoint.com/genetic_algorithms/genetic_algorithms_crossover.htm\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
