{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutations in genetic algorithms\n",
    "\n",
    "**Mutation** is the process of randomly altering some symbols in a solution's string. Mutation can be performed with different probabilities, such as fixed, adaptive, etc. <span style=\"color:#009988\">The idea is to introduce some diversity and variation in the population, which may help to escape from local optima and explore new regions of the search space.</span>\n",
    "\n",
    "Large initial mutation rate (probability that mutation occur in an individual) favors exploration due to causing large jumps in search space. As mutation rate decreases, exploitation is favored. <span style=\"color:#998800\">The probability of mutation could be proportional to the fitness of the individual, or we can simply decrease the mutation probability over time.</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/mutation.png\"  alt='missing' width=\"400\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Contents**\n",
    "\n",
    "<a href='#chrom-def'>01. Chromosome Definition</a>  \n",
    " \n",
    "<a href='#bit-flip'>02. Bit Flip Mutation</a>\n",
    "\n",
    "<a href='#random-resetting'>03. Random Resetting Mutation</a>\n",
    "\n",
    "<a href='#swap'>04. Swap Mutation</a>\n",
    "\n",
    "<a href='#scramble'>05. Scramble Mutation or Partial Shuffle Mutation (PSM)</a>\n",
    "\n",
    "<a href='#inversion'>06. Inversion Mutation or Reverse Sequence Mutation (RSM)</a>\n",
    "\n",
    "<a href='#inorder'>07. Inorder Mutation</a>\n",
    "\n",
    "<a href='#central-inversion'>08. Central Inversion Mutation (CIM)</a>\n",
    "\n",
    "<a href='#throas'>09. Throas Mutation</a>\n",
    "\n",
    "<a href='#thrors'>10. Thrors Mutation</a>\n",
    "\n",
    "<a href='#distance-based'>11. Distance-Based Mutation (DMO)</a>\n",
    "\n",
    "<a href='#displacement'>12. Displacement Mutation</a>\n",
    "\n",
    "<a href='#insertion'>13. Insertion Mutation</a>\n",
    "\n",
    "<a href='#displaced-inversion'>14. Displaced Inversion Mutation</a>\n",
    "\n",
    "<a href='#creep'>15. Creep Mutation</a>\n",
    "\n",
    "<a href='#uniform-random'>16. Uniform Random Mutation</a>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='chrom-def'>Chromosome Definition</a>\n",
    "\n",
    "First, we need to implement the chromosome class, a function to calculate the fitness of individuals (chromosomes) in the population, and some helper functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "class Chromosome():\n",
    "    \"\"\"\n",
    "    Description of class `Chromosome`:\n",
    "    This class represents a simple chromosome. In the method describe, a simple description\n",
    "    of the chromosome is provided, when it is called. \n",
    "    \"\"\"\n",
    "    def __init__(self, genes, id_=None, fitness=-1):\n",
    "        self.id_ = id_\n",
    "        self.genes = genes\n",
    "        self.fitness = fitness       \n",
    "       \n",
    "    def describe(self): \n",
    "        \"\"\"\n",
    "        Prints the ID, fitness, and genes\n",
    "        \"\"\"\n",
    "        print(f\"ID=#{self.id_}, Fitness={self.fitness}, \\nGenes=\\n{self.genes}\")\n",
    " \n",
    "    def get_chrom_length(self): \n",
    "        \"\"\"\n",
    "        Returns the length of `self.genes`\n",
    "        \"\"\"\n",
    "        return len(self.genes)\n",
    "\n",
    "def find_chrom_type(chrom):\n",
    "    \"\"\"\n",
    "    This function, takes a chromosome and returns its type (binary, integer, \n",
    "    or a floating-point chromosome).\n",
    "    chrom: The chromoseme which its type is calculated and returned.\n",
    "    \"\"\" \n",
    "    # if a floating-point individual\n",
    "    if np.issubdtype(chrom.dtype, np.floating):\n",
    "        return \"float\"\n",
    "    # if a binary individual\n",
    "    elif np.array_equal(chrom, chrom.astype(bool)):\n",
    "        return \"binary\"\n",
    "    # if an integer individual\n",
    "    elif np.issubdtype(chrom.dtype, np.integer):\n",
    "        return \"integer\"\n",
    "    \n",
    "    return \"binary\"\n",
    "\n",
    "# Note that calculating the fitness of a chromosome depends on the problem at hand,\n",
    "# and what you see in this function is just for understanding the concept.\n",
    "def fitness_function(chrom): \n",
    "    \"\"\"\n",
    "    This function, takes a chromosome and returns a value as its fitness.\n",
    "    chrom: The chromoseme which its fitness is calculated and returned.\n",
    "    \"\"\"  \n",
    "    fitness = 0\n",
    "    chrom_type = find_chrom_type(chrom)\n",
    "    # if we have a binary chromosome (count the number of 1s)\n",
    "    if chrom_type == \"binary\":\n",
    "        return np.count_nonzero(chrom == 1)\n",
    "    # if we have an integer chromosome (count values between 0 and 4)\n",
    "    elif chrom_type == \"integer\":\n",
    "        return np.count_nonzero((chrom >= 0) & (chrom <= 4))\n",
    "    # if we have a floating-point chromosome (count values between 0 and 0.5) \n",
    "    elif chrom_type == \"float\":\n",
    "        return np.count_nonzero((chrom >= 0) & (chrom <= 0.5))\n",
    "    \n",
    "    return -1\n",
    "\n",
    "\n",
    "def sorted_random_numbers(point_nums, interval):\n",
    "    \"\"\"\n",
    "    Returns several sorted random integers in an interval.\n",
    "    point_nums: number of points to be selected.\n",
    "    interval: the interval which the numbers are selected.\n",
    "    \"\"\"    \n",
    "    points = []\n",
    "    for _ in range(point_nums):\n",
    "        num = np.random.randint(1, abs(interval))\n",
    "        while num in points:\n",
    "            num = np.random.randint(1, abs(interval))\n",
    "        points.append(num)\n",
    "    return sorted(points)\n",
    "\n",
    "\n",
    "def integer_chrom_generator(n):\n",
    "    \"\"\"\n",
    "    Produces chromosomes with integer values for each gene. (integer alleles)\n",
    "    n: The length of the chromosome to be produced \n",
    "    \"\"\"  \n",
    "    a = []\n",
    "    for _ in range(n):\n",
    "        rand = np.random.randint(0, n)\n",
    "        while rand in a:\n",
    "            rand = np.random.randint(0, n)\n",
    "        else:\n",
    "            a.append(rand)\n",
    "    return np.array(a)\n",
    "\n",
    "\n",
    "def binary_chrom_generator(n, zero_prob, one_prob): \n",
    "    \"\"\"\n",
    "    Produces chromosomes with binary values for each gene. (binary alleles)\n",
    "    n: The length of the chromosome to be produced \n",
    "    zero_prob: The probability for the allele to be zero\n",
    "    one_prob: The probability for the allele to be one\n",
    "    \"\"\"   \n",
    "    return np.random.choice([0, 1], size=n, p=[zero_prob, one_prob])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='bit-flip'>Bit Flip Mutation</a>\n",
    "\n",
    "In bit-flip mutation, one or several positions are randomly selected, and their values are flipped. In other words, in the selected positions, if the value of the gene is 1, it will become 0; and if the value of the gene is 0, it will become 1."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/bit_flip_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 1 1 0 1 0 0 1 0]\n",
      "\n",
      "points: [3, 7]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[1 1 1 1 1 0 0 0 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def bit_flip_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Bit flip mutation on it. \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "    \n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        point_nums = np.random.randint(1, 4)\n",
    "        points = sorted_random_numbers(point_nums, chrom_length)\n",
    "        print(\"\\npoints:\", points)\n",
    "\n",
    "        for i in range(point_nums): #performing XOR, to flip the bits in the chosen points\n",
    "            parent_one.genes[points[i]] = parent_one.genes[points[i]] ^ 1\n",
    "        child_one.genes = parent_one.genes\n",
    "\n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 1, 1, 0, 1, 0, 0, 1, 0])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = bit_flip_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='random-resetting'>Random Resetting Mutation</a>\n",
    "\n",
    "In random resetting mutation, one or several positions are randomly selected, and a value is determined for each of those positions, from the allowed or valid domain. A domain could be the positive integers."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/random_resetting_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=125.2, \n",
      "Genes=\n",
      "[1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "points: [4, 7]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 2 3 4 7 6 7 9 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def random_resetting_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Random resetting mutation on it. \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "    parent_one = Chromosome(genes= np.array([1, 2, 3, 4, 5, 6, 7, 8, 9]),id_=0,fitness = 125.2)   \n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        point_nums = np.random.randint(1, 4)\n",
    "        points = sorted_random_numbers(point_nums, chrom_length)\n",
    "        print(\"\\npoints:\", points)\n",
    "\n",
    "        for i in range(len(points)): \n",
    "            parent_one.genes[points[i]] = np.random.randint(1, 10) #one of the admissible values\n",
    "        child_one.genes = parent_one.genes \n",
    "\n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 1, 1, 0, 1, 0, 0, 1, 0])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = random_resetting_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='swap'>Swap Mutation</a>\n",
    "In swap mutation, two genes are randomly selected, and the value of those two genes is swapped."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/swap_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "points: [2, 6]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 2 7 4 5 6 3 8 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def swap_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Swap mutation on it. \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "\n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        points = sorted_random_numbers(2, chrom_length)\n",
    "        print(\"\\npoints:\", points)    \n",
    "        parent_one.genes[points[0]], parent_one.genes[points[1]] = parent_one.genes[points[1]], parent_one.genes[points[0]] \n",
    "        child_one.genes = parent_one.genes \n",
    "\n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = swap_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='scramble'>Scramble Mutation or Partial Shuffle Mutation (PSM)</a>\n",
    "\n",
    "In scramble mutation, a substring of the chromosome is selected and the value of its genes are shuffled."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/scramble_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 4 2 7 5 3 6 8 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def scramble_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Scramble mutation on it. \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "\n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        points = sorted_random_numbers(2, chrom_length)   \n",
    "        permute = []\n",
    "        child_one.genes = parent_one.genes\n",
    "        for i in range(points[0], points[1]):\n",
    "            permute.append(parent_one.genes[i])\n",
    "        np.random.shuffle(permute)\n",
    "        k = 0\n",
    "        for i in range(points[0], points[1]):\n",
    "            child_one.genes[i] = permute[k]\n",
    "            k+=1      \n",
    "        \n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = scramble_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='inversion'>Inversion Mutation or Reverse Sequence Mutation (RSM)</a>\n",
    "\n",
    "In inversion mutation, a substring of the chromosome is selected and is inversed and replaced."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/inversion_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 8 7 6 5 4 3 2 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def inversion_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Inversion mutation on it. \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "\n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        points = sorted_random_numbers(2, chrom_length)  \n",
    "        reverse_list = []\n",
    "        child_one.genes = parent_one.genes\n",
    "        for i in range(points[0], points[1]):\n",
    "            reverse_list.append(parent_one.genes[i])\n",
    "        reverse_list.reverse()\n",
    "        # print(reverse_list)\n",
    "        k = 0\n",
    "        for i in range(points[0], points[1]):\n",
    "            child_one.genes[i] = reverse_list[k]\n",
    "            k+=1      \n",
    "         \n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = inversion_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='inorder'>Inorder Mutation</a>\n",
    "\n",
    "In Inorder mutation, several random positions are selected, and for each position, a uniform random real number `w` is produced. If `w < 0.5` for that position, the value of the gene in that position is flipped. Otherwise, the value of the gene will remain unchanged."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/inorder_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 0 1 1 0 1 1 0]\n",
      "\n",
      "points: [2, 5, 6, 7]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 1 1 1 0 0 1 0]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def inorder_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Inorder mutation on it. \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "    \n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        point_nums = np.random.randint(1, 5) #selecting one to four elements \n",
    "        points = sorted_random_numbers(point_nums, chrom_length)\n",
    "        child_one.genes = parent_one.genes\n",
    "        print(\"\\npoints:\", points)    \n",
    "        for i in range(point_nums):\n",
    "            w = np.random.uniform(low=0, high=1)\n",
    "            if w < 0.5:\n",
    "                child_one.genes[points[i]] = child_one.genes[points[i]] ^ 1 #flipping the bit\n",
    "\n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 0, 1, 1, 0, 1, 1, 0])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = inorder_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='center-inversion'>Center Inversion Mutation (CIM)</a>\n",
    "\n",
    "In center inversion mutation, we divide the chromosome into two parts; then we reverse each part and transfer the result to the offspring."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/center_inversion_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[4 3 2 1 9 8 7 6 5]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def center_inversion_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Center inversion mutation on it. \n",
    "    Length of the chromosome in this mutation should be more than 4.\n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "\n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        reverse_list1 = []\n",
    "        reverse_list2 = []\n",
    "        point = np.random.randint(2,chrom_length-1) #break point  \n",
    "        child_one.genes = parent_one.genes\n",
    "        for i in range(chrom_length):\n",
    "            if i < point:\n",
    "                reverse_list1.append(parent_one.genes[i])\n",
    "            else:\n",
    "                reverse_list2.append(parent_one.genes[i])\n",
    "        reverse_list1.reverse()\n",
    "        reverse_list2.reverse()\n",
    "\n",
    "        for i in range(point): #Filling the first half\n",
    "            child_one.genes[i] = reverse_list1[i]\n",
    "        k = 0\n",
    "        for i in range(point, chrom_length): #Filling the first half\n",
    "            child_one.genes[i] = reverse_list2[k]\n",
    "            k+=1\n",
    "                \n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "    \n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "\n",
    "    return child_one\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = center_inversion_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='throas'>Throas Mutation</a>\n",
    "In throas mutation, we select three positions in a sequence and do the following. <br>\n",
    "> (1) Place the last element in the first position in the child.<br>\n",
    "> (2) Place the second element in the last position in the child.<br>\n",
    "> (3) Place the first element in the second position in the child.\n",
    "\n",
    "All the other elements are directly transferred from parent-one to child-one."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/throas_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 2 5 3 4 6 7 8 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def throas_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Throas mutation on it. \n",
    "    Last ==> First\n",
    "    Second ==> Last\n",
    "    First ==> Second \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "\n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        interval = []\n",
    "        point1 = np.random.randint(0,chrom_length-3)\n",
    "        point2 = point1 + 3\n",
    "        child_one.genes = parent_one.genes\n",
    "        for i in range(point1, point2):\n",
    "            interval.append(child_one.genes[i])\n",
    "        # print(interval)\n",
    "        interval_copy = interval.copy()\n",
    "        interval[0] = interval_copy[-1] \n",
    "        interval[-1] = interval_copy[1] \n",
    "        interval[1] = interval_copy[0] \n",
    "        # print(interval)\n",
    "        k = 0\n",
    "        for i in range(point1, point2):\n",
    "            child_one.genes[i] = interval[k]\n",
    "            k+=1\n",
    "\n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = throas_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='thrors'>Thrors Mutation</a>\n",
    "\n",
    "In thrors mutation, we select three positions that are not necessarily in a sequence and do the following.<br>\n",
    "> (1) Place the last element in the first position in the child. <br>\n",
    "> (2) Place the second element in the last position in the child. <br>\n",
    "> (3) Place the first element in the second position in the child.\n",
    "\n",
    "All the other elements are directly transferred from parent-one to child-one."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/thrors_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=4, \n",
      "Genes=\n",
      "[1 8 2 4 5 6 7 3 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def thrors_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Thrors mutation on it. \n",
    "    Last ==> First\n",
    "    Second ==> Last\n",
    "    First ==> Second \n",
    "    In compare to Throas mutation, in Thrors mutation, the chosen elements are not necessarily \n",
    "    successive. \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "\n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        elements = []\n",
    "        points = sorted_random_numbers(3, chrom_length)\n",
    "        # print(points)    \n",
    "        child_one.genes = parent_one.genes\n",
    "\n",
    "        for i in range(len(points)): #selecting the elements\n",
    "            elements.append(parent_one.genes[points[i]]) \n",
    "        # print(elements)\n",
    "        elements_copy = elements.copy()\n",
    "        elements[0] = elements_copy[-1] \n",
    "        elements[-1] = elements_copy[1] \n",
    "        elements[1] = elements_copy[0] \n",
    "        # print(elements)\n",
    "        k = 0\n",
    "        for i in range(chrom_length):\n",
    "            if i in points:\n",
    "                child_one.genes[i] = elements[k]\n",
    "                k+=1\n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = thrors_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='distance-based'>Distance-Based Mutation (DMO)</a>\n",
    "\n",
    "In distance-based mutation, we select a random position and find the elements at a certain distance around the selected position. After finding the selected elements, we shuffle the whole interval.\n",
    "\n",
    "To calculate the selected elements, we define two variables: the variable `g` which indicates the randomly selected position, and the variable `distance` to indicate which elements on either side of the randomly selected element will be selected."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/distance_based_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "g = 4, distance = 3\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 5 2 1 7 3 4 6 8 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def distance_based_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Distance based mutation on it.\n",
    "    In this method, g is the position for the chosen gene. d is the distance from that position. \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "\n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        interval = []\n",
    "        d = np.random.randint(1, 4)\n",
    "        g = np.random.randint(d,chrom_length-d)\n",
    "        print(\"\\ng = {}, distance = {}\".format(g, d))\n",
    "        child_one.genes = parent_one.genes\n",
    "        for i in range(g - d, g + d + 1): #element in position g and d elements in each direction\n",
    "            interval.append(parent_one.genes[i]) \n",
    "        np.random.shuffle(interval)\n",
    "        k = 0\n",
    "        for i in range(g - d, g + d + 1): #Replacing elements\n",
    "            child_one.genes[i] = interval[k]\n",
    "            k+=1\n",
    "\n",
    "\n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = distance_based_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='displacement'>Displacement Mutation</a>\n",
    "\n",
    "In displacement mutation, we select a random substring from the chromosome, and paste it in another position, without changing the sequence of other elements."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/displacement_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 7 8 9 1 2 3 4 5 6]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def displacement_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Displacement mutation on it. \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "\n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        interval = []\n",
    "        points = sorted_random_numbers(2, chrom_length)\n",
    "        # print(points)    \n",
    "        duplicate = parent_one.genes.tolist().copy() \n",
    "        for i in range(points[0], points[1]):\n",
    "            interval.append(parent_one.genes[i])\n",
    "        # print(interval)\n",
    "        for i in range(chrom_length):\n",
    "            if parent_one.genes[i] in interval:\n",
    "                duplicate.remove(parent_one.genes[i])\n",
    "   \n",
    "        # print(duplicate)\n",
    "        lists = []\n",
    "        for i in range(len(duplicate) + 1):\n",
    "            a = []\n",
    "            lists.append(a)\n",
    "        position = np.random.randint(0, len(duplicate) + 1) #the position of the paste\n",
    "        # print(position)\n",
    "        for i in range(len(interval)):\n",
    "            lists[position].append(interval[i])\n",
    "        k = 0\n",
    "        for i in range(len(lists)):\n",
    "            if not lists[i]: #if lists[i] is empty\n",
    "                lists[i].append(duplicate[k])\n",
    "                k+=1\n",
    "        # print(lists)\n",
    "        flat_list = [item for sublist in lists for item in sublist]\n",
    "        child_one.genes = np.array(flat_list)\n",
    "\n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "    \n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "\n",
    "    return child_one\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = displacement_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='insertion'>Insertion Mutation</a>\n",
    "\n",
    "In insertion mutation, we select a random gene in the chromosome and we paste that gene in another position, without changing the sequence of other elements. The main difference between insertion mutation and displacement mutation is that in insertion mutation, we change the position of only one element."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/insertion_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 2 3 4 1 5 6 7 8 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def insertion_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Insertion mutation on it. \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "\n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        point = np.random.randint(0, chrom_length) #picking one gene\n",
    "        # print(point)    \n",
    "        duplicate = parent_one.genes.tolist().copy()         \n",
    "        duplicate.remove(parent_one.genes[point]) \n",
    "        # print(duplicate)\n",
    "        \n",
    "        lists = []\n",
    "        for i in range(len(duplicate) + 1):\n",
    "            a = []\n",
    "            lists.append(a)\n",
    "        position = np.random.randint(0, len(duplicate) + 1) #the position of the paste\n",
    "        # print(position)\n",
    "        lists[position].append(parent_one.genes[point])\n",
    "        k = 0\n",
    "        for i in range(len(lists)):\n",
    "            if not lists[i]: #if lists[i] is empty\n",
    "                lists[i].append(duplicate[k])\n",
    "                k+=1\n",
    "        # print(lists)\n",
    "        flat_list = [item for sublist in lists for item in sublist]\n",
    "        child_one.genes = np.array(flat_list)\n",
    "\n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = insertion_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='displaced-inversion'>Displaced Inversion Mutation</a>\n",
    "\n",
    "In displaced inversion mutation, we select a random substring from the chromosome, reverse that substring, and paste it in another position, without changing the sequence of other elements."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/displaced_inversion_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0 1 2 5 6 7 4 3 8 9]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def displaced_inversion_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Displaced inversion mutation on it. \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "\n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        interval = []\n",
    "        points = sorted_random_numbers(2, chrom_length)\n",
    "        # print(points)    \n",
    "        duplicate = parent_one.genes.tolist().copy() \n",
    "        for i in range(points[0], points[1]):\n",
    "            interval.append(parent_one.genes[i])\n",
    "        # print(interval)\n",
    "        for i in range(chrom_length):\n",
    "            if parent_one.genes[i] in interval:\n",
    "                duplicate.remove(parent_one.genes[i])\n",
    "   \n",
    "        # print(duplicate)\n",
    "        lists = []\n",
    "        for i in range(len(duplicate) + 1):\n",
    "            a = []\n",
    "            lists.append(a)\n",
    "        position = np.random.randint(0, len(duplicate) + 1) #the position of the paste\n",
    "        # print(position)\n",
    "        interval.reverse()\n",
    "        for i in range(len(interval)):\n",
    "            lists[position].append(interval[i])\n",
    "\n",
    "        k = 0\n",
    "        for i in range(len(lists)):\n",
    "            if not lists[i]: #if lists[i] is empty\n",
    "                lists[i].append(duplicate[k])\n",
    "                k+=1\n",
    "        # print(lists)\n",
    "        flat_list = [item for sublist in lists for item in sublist]\n",
    "        child_one.genes = np.array(flat_list)\n",
    "\n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "chrom0 = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = displaced_inversion_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='creep'>Creep Mutation</a>\n",
    "\n",
    "**Creep mutation** is a type of mutation used in **real-value encoding schemes** in genetic algorithms. It involves the addition of a small random vector to the elements of a chromosome or an element-wise approach involving a probability test on each element to determine whether to perform the mutation. \n",
    "\n",
    "In the following implementation, we choose a random position in the parent chromosome and replace its value with a random value chosen from the valid domain (in our case, in `[0, 1]`). \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/creep_mutation.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.121 0.152 0.231 0.143 0.732 0.315 0.434 0.633]\n",
      "\n",
      "Chosen point: 5\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=5, \n",
      "Genes=\n",
      "[0.121 0.152 0.231 0.143 0.732 0.703 0.434 0.633]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def creep_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    This function takes one chromosome, and performs Creep mutation on it. \n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    child_one = Chromosome(genes= np.array([0.0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        point = np.random.randint(0, chrom_length)\n",
    "        print(\"\\nChosen point:\", point)    \n",
    "        child_one.genes = parent_one.genes \n",
    "        child_one.genes[point] = round(np.random.uniform(low=0.1, high=0.95), 3)\n",
    "\n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "    \n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0.121, 0.152, 0.231, 0.143, 0.732, 0.315, 0.434, 0.633])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "MUTATION = creep_mutation(parent_one, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='uniform-random'>Uniform Random Mutation</a>\n",
    "\n",
    "In **Uniform Random Mutation**, a new individual is created by randomly generating a value for each gene in the chromosome. The new individual is then subjected to uniform crossover with the original individual. \n",
    "\n",
    "This mutation operator is used to introduce new genetic material into the population and to maintain genetic diversity. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <center> <img src=\"./pics/Mutation/uniform_random.png\"  alt='missing' width=\"450\"  ><center/>\n",
    "<figure/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Parent\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.121 0.152 0.231 0.143 0.732 0.315 0.434 0.633]\n",
      "ID=#0, Fitness=1, \n",
      "Genes=\n",
      "[0.904 0.903 0.576 0.759 0.341 0.612 0.721 0.934]\n",
      "\n",
      "Mask vector is: [0, 1, 1, 1, 0, 1, 1, 1]\n",
      "\n",
      "Children\n",
      "=================================================\n",
      "ID=#0, Fitness=6, \n",
      "Genes=\n",
      "[0.904 0.152 0.231 0.143 0.341 0.315 0.434 0.633]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "def claculate_mask(chrom_size, px):\n",
    "    \"\"\"\n",
    "    This method, performs the mask calculation for the Uniform crossover.\n",
    "    chrom_size: The mask has the same size of the chromosome.\n",
    "    px: A control parameter for the number of ones in the mask vector.\n",
    "    \"\"\"  \n",
    "    mask = [0]*chrom_size\n",
    "    for i in range(chrom_size):\n",
    "        prob = np.random.rand()\n",
    "        if prob <= px:\n",
    "            mask[i]=1\n",
    "    return mask\n",
    "\n",
    "\n",
    "def uniform_random_mutation(parent_one, pm):\n",
    "    \"\"\"\n",
    "    parent_one: The parent\n",
    "    pm: The probability of mutation\n",
    "    \"\"\" \n",
    "\n",
    "    print(\"\\nParent\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(parent_one)\n",
    "    chrom_length = Chromosome.get_chrom_length(parent_one)\n",
    "    chrom_type = find_chrom_type(parent_one.genes)\n",
    "\n",
    "    \n",
    "    if chrom_type == \"float\":\n",
    "        random_genes = np.round(np.random.uniform(0.0, 1.0, chrom_length), 3)\n",
    "        parent_two = Chromosome(genes= random_genes, id_=0, fitness = fitness_function(random_genes))\n",
    "        child_one = Chromosome(genes= np.array([0.0]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    \n",
    "    elif chrom_type == \"integer\":\n",
    "        random_genes = integer_chrom_generator(chrom_length)\n",
    "        parent_two = Chromosome(genes= random_genes, id_=0, fitness = fitness_function(random_genes))\n",
    "        child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "    \n",
    "    else: # binary individual\n",
    "        random_genes = np.array(binary_chrom_generator(chrom_length, 0.5, 0.5))\n",
    "        parent_two = Chromosome(genes= random_genes, id_=0, fitness = fitness_function(random_genes))\n",
    "        child_one = Chromosome(genes= np.array([-1]*chrom_length),id_=0,fitness = 125.2)   \n",
    "\n",
    "    # parent_two is a randomly generated chromosome (not in the population)\n",
    "    Chromosome.describe(parent_two)\n",
    "\n",
    "    if np.random.rand() < pm:  # if pm is greater than random number\n",
    "        mask = claculate_mask(chrom_length, np.random.uniform(low=0.2, high=0.85))\n",
    "        print(\"\\nMask vector is:\", mask)\n",
    "        for i in range(chrom_length):\n",
    "            if mask[i]==1:\n",
    "                child_one.genes[i] = parent_one.genes[i]\n",
    "            else:\n",
    "                child_one.genes[i] = parent_two.genes[i]         \n",
    "\n",
    "    else:  # if pm is less than random number then don't make any change\n",
    "        child_one = deepcopy(parent_one)\n",
    "\n",
    "    # updating fitnesses for children\n",
    "    child_one.fitness = fitness_function(child_one.genes)\n",
    "    \n",
    "    return child_one\n",
    "\n",
    "\n",
    "\n",
    "chrom0 = np.array([0.121, 0.152, 0.231, 0.143, 0.732, 0.315, 0.434, 0.633])\n",
    "chrom1 = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])\n",
    "chrom2 = np.array([0, 1, 0, 0, 1, 0, 0, 1, 1, 0])\n",
    "parent_one = Chromosome(genes= chrom0, id_=0, fitness = fitness_function(chrom0))   \n",
    "parent_two = Chromosome(genes= chrom1, id_=1, fitness = fitness_function(chrom1))   \n",
    "parent_three = Chromosome(genes= chrom2, id_=2, fitness = fitness_function(chrom2))   \n",
    "\n",
    "# You can try this mutation with any kind of representation (e.g., float, binary, integer, etc.)\n",
    "MUTATION = uniform_random_mutation(parent_one, 1)\n",
    "# MUTATION = uniform_random_mutation(parent_two, 1)\n",
    "# MUTATION = uniform_random_mutation(parent_three, 1)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    print(\"\\nChildren\")\n",
    "    print(\"=================================================\")\n",
    "    Chromosome.describe(MUTATION)  #one child\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References \n",
    "\n",
    "1. Engelbrecht, A. P. (2007). Computational intelligence: An introduction (2nd ed.). John Wiley & Sons.\n",
    "\n",
    "2. Bing, M. (2023). Recombination in genetic algorithms [Graphic art]. Image Creator from Microsoft Designer.\n",
    "\n",
    "3. Optimization and Parameter Estimation, Genetic Algorithms. (n.d.). In SpringerLink. Retrieved December 27, 2023, from https://link.springer.com/referenceworkentry/10.1007/978-1-4419-9863-7_291\n",
    "\n",
    "4. Zhang, Y., & Wang, X. (2021). A Method Enabling Comprehensive Isolation of Arabidopsis Root Cell Types for Transcriptional Analysis. Frontiers in Plant Science, 12, 646404. https://doi.org/10.3389/fpls.2021.646404\n",
    "\n",
    "5. Creepy Mutant. (n.d.). In The Forest Wiki. Retrieved December 27, 2023, from https://theforest.fandom.com/wiki/Creepy_Mutant\n",
    "\n",
    "6. Koza, J. R. (1992). Exact GP schema theory for headless chicken crossover and subtree mutation. In Proceedings of the 4th International Conference on Genetic Algorithms (pp. 78-85). IEEE. https://doi.org/10.1109/ICGA.1991.576974\n",
    "\n",
    "7. Gorman, J. (2015, March 11). Why Transylvanian Chickens Have Naked Necks. National Geographic. Retrieved December 27, 2023, from https://www.nationalgeographic.com/animals/article/110315-transylvanian-naked-neck-chicken-churkeys-turkens-science\n",
    "\n",
    "8. Kummer, C. (2014, August 27). Here's Why a Chicken Can Live Without Its Head. Modern Farmer. Retrieved December 27, 2023, from https://modernfarmer.com/2014/08/heres-chicken-can-live-without-head/\n",
    "\n",
    "9. Li, Y., Wu, F., & Zhang, Y. (2020). The genetic basis and robustness of naked neck mutation in chicken: A comprehensive analysis. Tropical Animal Health and Production, 52(6), 3049-3056. https://doi.org/10.1007/s11250-020-02505-1\n",
    "\n",
    "10. IEEE Xplore Digital Library. (n.d.). Retrieved December 27, 2023, from https://ieeexplore.ieee.org/servlet/opac?punumber=7440.\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
